{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0d6e2815-39ac-43e7-9da2-875fa6b31bce",
   "metadata": {
    "tags": []
   },
   "source": [
    "# ReviewerCatalog"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "403e60ad-39cf-4448-bb54-03f1ee77c7b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67c55c3e-ca0d-460d-a0ed-677779f069e2",
   "metadata": {},
   "source": [
    "# Registry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "e9e0174d-6485-4f3f-9e31-eb06530cfcf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from ReviewerCatalog import display_catalog_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "d50e60cb-3bb5-45f0-969f-47ab67aee4d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_fd7f9 th {\n",
       "  vertical-align: top;\n",
       "}\n",
       "#T_fd7f9 .row0 {\n",
       "  border-top: 1px solid black;\n",
       "}\n",
       "#T_fd7f9 .row15 {\n",
       "  border-top: 1px solid black;\n",
       "}\n",
       "#T_fd7f9 .row13 {\n",
       "  border-top: 1px solid black;\n",
       "}\n",
       "#T_fd7f9 .row23 {\n",
       "  border-top: 1px solid black;\n",
       "}\n",
       "#T_fd7f9 .row7 {\n",
       "  border-top: 1px solid black;\n",
       "}\n",
       "#T_fd7f9_row0_col0, #T_fd7f9_row0_col1, #T_fd7f9_row1_col0, #T_fd7f9_row1_col1, #T_fd7f9_row2_col0, #T_fd7f9_row2_col1, #T_fd7f9_row3_col0, #T_fd7f9_row3_col1, #T_fd7f9_row4_col0, #T_fd7f9_row4_col1, #T_fd7f9_row5_col0, #T_fd7f9_row5_col1, #T_fd7f9_row6_col0, #T_fd7f9_row6_col1, #T_fd7f9_row7_col0, #T_fd7f9_row7_col1, #T_fd7f9_row8_col0, #T_fd7f9_row8_col1, #T_fd7f9_row9_col0, #T_fd7f9_row9_col1, #T_fd7f9_row10_col0, #T_fd7f9_row10_col1, #T_fd7f9_row11_col0, #T_fd7f9_row11_col1, #T_fd7f9_row12_col0, #T_fd7f9_row12_col1, #T_fd7f9_row13_col0, #T_fd7f9_row13_col1, #T_fd7f9_row14_col0, #T_fd7f9_row14_col1, #T_fd7f9_row15_col0, #T_fd7f9_row15_col1, #T_fd7f9_row16_col0, #T_fd7f9_row16_col1, #T_fd7f9_row17_col0, #T_fd7f9_row17_col1, #T_fd7f9_row18_col0, #T_fd7f9_row18_col1, #T_fd7f9_row19_col0, #T_fd7f9_row19_col1, #T_fd7f9_row20_col0, #T_fd7f9_row20_col1, #T_fd7f9_row21_col0, #T_fd7f9_row21_col1, #T_fd7f9_row22_col0, #T_fd7f9_row22_col1, #T_fd7f9_row23_col0, #T_fd7f9_row23_col1, #T_fd7f9_row24_col0, #T_fd7f9_row24_col1 {\n",
       "  background-color: white;\n",
       "  vertical-align: text-top;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_fd7f9\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank\" >&nbsp;</th>\n",
       "      <th class=\"blank\" >&nbsp;</th>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_fd7f9_level0_col0\" class=\"col_heading level0 col0\" >url</th>\n",
       "      <th id=\"T_fd7f9_level0_col1\" class=\"col_heading level0 col1\" >Description</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th class=\"index_name level0\" >Repo</th>\n",
       "      <th class=\"index_name level1\" >Type</th>\n",
       "      <th class=\"index_name level2\" >Name</th>\n",
       "      <th class=\"blank col0\" >&nbsp;</th>\n",
       "      <th class=\"blank col1\" >&nbsp;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level0_row0\" class=\"row_heading level0 row0\" rowspan=\"7\">getzlab/AnnoMate</th>\n",
       "      <th id=\"T_fd7f9_level1_row0\" class=\"row_heading level1 row0\" >Repository</th>\n",
       "      <th id=\"T_fd7f9_level2_row0\" class=\"row_heading level2 row0\" >getzlab/AnnoMate</th>\n",
       "      <td id=\"T_fd7f9_row0_col0\" class=\"data row0 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/AnnoMate\">AnnoMate</a></td>\n",
       "      <td id=\"T_fd7f9_row0_col1\" class=\"data row0 col1\" >Review anything (purities, mutations, etc) within a jupyter notebook with plotly dash and jupyter widgets</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level1_row1\" class=\"row_heading level1 row1\" rowspan=\"5\">AppComponent</th>\n",
       "      <th id=\"T_fd7f9_level2_row1\" class=\"row_heading level2 row1\" >CNVPlotComponent</th>\n",
       "      <td id=\"T_fd7f9_row1_col0\" class=\"data row1 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/AnnoMate/blob/master/AnnoMate/AppComponents/CNVPlotComponent.py\">CNVPlotComponent.py</a></td>\n",
       "      <td id=\"T_fd7f9_row1_col1\" class=\"data row1 col1\" >CNVPlotComponent.py module\n",
       "\n",
       "Interactive CNV Plot with mutation multiplicity scatterplot\n",
       "\n",
       "Mutation scatter interactive with mutation table\n",
       "\n",
       "</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level2_row2\" class=\"row_heading level2 row2\" >DataTableComponents</th>\n",
       "      <td id=\"T_fd7f9_row2_col0\" class=\"data row2 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/AnnoMate/blob/master/AnnoMate/AppComponents/DataTableComponents.py\">DataTableComponents.py</a></td>\n",
       "      <td id=\"T_fd7f9_row2_col1\" class=\"data row2 col1\" >\n",
       "DataTableComponents module contains methods to generate components for displaying table information\n",
       "</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level2_row3\" class=\"row_heading level2 row3\" >MutationTableComponent</th>\n",
       "      <td id=\"T_fd7f9_row3_col0\" class=\"data row3 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/AnnoMate/blob/master/AnnoMate/AppComponents/MutationTableComponent.py\">MutationTableComponent.py</a></td>\n",
       "      <td id=\"T_fd7f9_row3_col1\" class=\"data row3 col1\" >MutationTableComponent.py module\n",
       "\n",
       "Interactive Mutation Table with column selection, sorting, selecting, and filtering\n",
       "\n",
       "</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level2_row4\" class=\"row_heading level2 row4\" >PhylogicComponents</th>\n",
       "      <td id=\"T_fd7f9_row4_col0\" class=\"data row4 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/AnnoMate/blob/master/AnnoMate/AppComponents/PhylogicComponents.py\">PhylogicComponents.py</a></td>\n",
       "      <td id=\"T_fd7f9_row4_col1\" class=\"data row4 col1\" >PhylogicComponents.py module\n",
       "\n",
       "Phylogic CCF Plot and Trees implemented in the PatientReviewer and PhylogicReviewer\n",
       "\n",
       "Phylogic PMF Plot implemented in the PhylogicReviewer\n",
       "\n",
       "</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level2_row5\" class=\"row_heading level2 row5\" >utils</th>\n",
       "      <td id=\"T_fd7f9_row5_col0\" class=\"data row5 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/AnnoMate/blob/master/AnnoMate/AppComponents/utils.py\">utils.py</a></td>\n",
       "      <td id=\"T_fd7f9_row5_col1\" class=\"data row5 col1\" >Generates unique string for this mutation, including contig, start position, ref and alt alleles.\n",
       "\n",
       "    Does not include End Position, for this field is not present in mut_ccfs Phylogic output. However, specification of both the alt and ref alleles are enough to distinguish InDels.\n",
       "\n",
       "    :param row: pd.Series giving the data for one mutation from a maf or maf-like dataframe\n",
       "    :param chrom: the name of the contig/chromosome column/field; default: Chromosome\n",
       "    :param start_pos: the name of the start position column/field; default: Start_position\n",
       "    :param ref: the name of the reference allele column/field; default: Reference_Allele\n",
       "    :param alt: the name of the alternate allele column/field; default: Tumor_Seq_Allele\n",
       "    </td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level1_row6\" class=\"row_heading level1 row6\" >Reviewer</th>\n",
       "      <th id=\"T_fd7f9_level2_row6\" class=\"row_heading level2 row6\" >ExampleReviewer</th>\n",
       "      <td id=\"T_fd7f9_row6_col0\" class=\"data row6 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/AnnoMate/blob/master/AnnoMate/Reviewers/ExampleReviewer.py\">ExampleReviewer.py</a></td>\n",
       "      <td id=\"T_fd7f9_row6_col1\" class=\"data row6 col1\" >\n",
       "        Creates a dataframe with index corresponding to the primary dataframe's index\n",
       "\n",
       "        Parameters\n",
       "        ----------\n",
       "        sample_df: pd.DataFrame\n",
       "            dataframe with samples data and sample ids in the index\n",
       "        preprocessing_str: str\n",
       "            Example of preprocessing data\n",
       "        </td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level0_row7\" class=\"row_heading level0 row7\" rowspan=\"6\">getzlab/PurityReviewer</th>\n",
       "      <th id=\"T_fd7f9_level1_row7\" class=\"row_heading level1 row7\" >Repository</th>\n",
       "      <th id=\"T_fd7f9_level2_row7\" class=\"row_heading level2 row7\" >getzlab/PurityReviewer</th>\n",
       "      <td id=\"T_fd7f9_row7_col0\" class=\"data row7 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/PurityReviewer\">PurityReviewer</a></td>\n",
       "      <td id=\"T_fd7f9_row7_col1\" class=\"data row7 col1\" >Suite of purity reviewers and review components</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level1_row8\" class=\"row_heading level1 row8\" rowspan=\"3\">AppComponent</th>\n",
       "      <th id=\"T_fd7f9_level2_row8\" class=\"row_heading level2 row8\" >AbsoluteCustomSolutionComponent</th>\n",
       "      <td id=\"T_fd7f9_row8_col0\" class=\"data row8 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/PurityReviewer/blob/master/PurityReviewer/AppComponents/AbsoluteCustomSolutionComponent.py\">AbsoluteCustomSolutionComponent.py</a></td>\n",
       "      <td id=\"T_fd7f9_row8_col1\" class=\"data row8 col1\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level2_row9\" class=\"row_heading level2 row9\" >AbsoluteSolutionsReportComponent</th>\n",
       "      <td id=\"T_fd7f9_row9_col0\" class=\"data row9 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/PurityReviewer/blob/master/PurityReviewer/AppComponents/AbsoluteSolutionsReportComponent.py\">AbsoluteSolutionsReportComponent.py</a></td>\n",
       "      <td id=\"T_fd7f9_row9_col1\" class=\"data row9 col1\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level2_row10\" class=\"row_heading level2 row10\" >utils</th>\n",
       "      <td id=\"T_fd7f9_row10_col0\" class=\"data row10 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/PurityReviewer/blob/master/PurityReviewer/AppComponents/utils.py\">utils.py</a></td>\n",
       "      <td id=\"T_fd7f9_row10_col1\" class=\"data row10 col1\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level1_row11\" class=\"row_heading level1 row11\" rowspan=\"2\">Reviewer</th>\n",
       "      <th id=\"T_fd7f9_level2_row11\" class=\"row_heading level2 row11\" >ManualPurityReviewer</th>\n",
       "      <td id=\"T_fd7f9_row11_col0\" class=\"data row11 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/PurityReviewer/blob/master/PurityReviewer/Reviewers/ManualPurityReviewer.py\">ManualPurityReviewer.py</a></td>\n",
       "      <td id=\"T_fd7f9_row11_col1\" class=\"data row11 col1\" >Generate data for PurityReviewer object\n",
       "\n",
       "        Returns\n",
       "        -------\n",
       "        GenericData\n",
       "            Data object that contains only one dataframe\n",
       "        </td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level2_row12\" class=\"row_heading level2 row12\" >MatchedPurityReviewer</th>\n",
       "      <td id=\"T_fd7f9_row12_col0\" class=\"data row12 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/PurityReviewer/blob/master/PurityReviewer/Reviewers/MatchedPurityReviewer.py\">MatchedPurityReviewer.py</a></td>\n",
       "      <td id=\"T_fd7f9_row12_col1\" class=\"data row12 col1\" >Generate data for PurityReviewer object\n",
       "\n",
       "        Returns\n",
       "        -------\n",
       "        GenericData\n",
       "            Data object that contains only one dataframe\n",
       "        </td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level0_row13\" class=\"row_heading level0 row13\" rowspan=\"2\">getzlab/PatientReviewer</th>\n",
       "      <th id=\"T_fd7f9_level1_row13\" class=\"row_heading level1 row13\" >Repository</th>\n",
       "      <th id=\"T_fd7f9_level2_row13\" class=\"row_heading level2 row13\" >getzlab/PatientReviewer</th>\n",
       "      <td id=\"T_fd7f9_row13_col0\" class=\"data row13 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/PatientReviewer\">PatientReviewer</a></td>\n",
       "      <td id=\"T_fd7f9_row13_col1\" class=\"data row13 col1\" >Explore integrated data on the patient level interactively in a Dash App, powered by JupyterReviewer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level1_row14\" class=\"row_heading level1 row14\" >Reviewer</th>\n",
       "      <th id=\"T_fd7f9_level2_row14\" class=\"row_heading level2 row14\" >PatientReviewer</th>\n",
       "      <td id=\"T_fd7f9_row14_col0\" class=\"data row14 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/PatientReviewer/blob/master/PatientReviewer/Reviewers/PatientReviewer.py\">PatientReviewer.py</a></td>\n",
       "      <td id=\"T_fd7f9_row14_col1\" class=\"data row14 col1\" >PatientReviewer.py module\n",
       "\n",
       "Interactive dashboard for reviewing and annotating data on a patient-by-patient basis\n",
       "Includes app layout and callback functionality\n",
       "\n",
       "Run by the user with a Jupyter Notebook: UserPatientReviewer.ipynb\n",
       "\n",
       "</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level0_row15\" class=\"row_heading level0 row15\" rowspan=\"8\">getzlab/MutationReviewer</th>\n",
       "      <th id=\"T_fd7f9_level1_row15\" class=\"row_heading level1 row15\" >Repository</th>\n",
       "      <th id=\"T_fd7f9_level2_row15\" class=\"row_heading level2 row15\" >getzlab/MutationReviewer</th>\n",
       "      <td id=\"T_fd7f9_row15_col0\" class=\"data row15 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/MutationReviewer\">MutationReviewer</a></td>\n",
       "      <td id=\"T_fd7f9_row15_col1\" class=\"data row15 col1\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level1_row16\" class=\"row_heading level1 row16\" rowspan=\"5\">AppComponent</th>\n",
       "      <th id=\"T_fd7f9_level2_row16\" class=\"row_heading level2 row16\" >BamTableComponent</th>\n",
       "      <td id=\"T_fd7f9_row16_col0\" class=\"data row16 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/MutationReviewer/blob/master/MutationReviewer/AppComponents/BamTableComponent.py\">BamTableComponent.py</a></td>\n",
       "      <td id=\"T_fd7f9_row16_col1\" class=\"data row16 col1\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level2_row17\" class=\"row_heading level2 row17\" >IGVJSComponent</th>\n",
       "      <td id=\"T_fd7f9_row17_col0\" class=\"data row17 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/MutationReviewer/blob/master/MutationReviewer/AppComponents/IGVJSComponent.py\">IGVJSComponent.py</a></td>\n",
       "      <td id=\"T_fd7f9_row17_col1\" class=\"data row17 col1\" >\n",
       "    Callback function to generate an IGV.js window centered around the locus/loci of interest \n",
       "    and specified bams loaded when the Update Tracks button is clicked.\n",
       "    \n",
       "    Parameters\n",
       "    ------\n",
       "    update_tracks_n_clicks: State\n",
       "        Dash.State of the number of times a button was clicked\n",
       "        \n",
       "    bam_table: State\n",
       "        Dash.State object referencing a state of a dash component containing \n",
       "        a table with the bam files\n",
       "        \n",
       "    bam_table_selected_rows: State, list\n",
       "        Dash.State object referencing a state of a dash component referencing \n",
       "        which rows are selected in a table with the bam files\n",
       "        \n",
       "    genome: string, default='hg19'\n",
       "            Name of genome to use in IGV.js\n",
       "        \n",
       "    track_height: int, default=400\n",
       "        Height to display each track in IGV.js mode\n",
       "\n",
       "    minimumBases: int, default=200\n",
       "        Minimum number of bases to display in a window in IGV.js mode\n",
       "        \n",
       "    gen_data_mut_index_name_func: func\n",
       "        Function used to parse the index to filter the mutation table\n",
       "        \n",
       "    Return\n",
       "    ------\n",
       "    \n",
       "    A dash_bio.IGV component\n",
       "        Contains the tracks specified to load and window centering around the locus (or loci) of interest\n",
       "    \n",
       "    </td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level2_row18\" class=\"row_heading level2 row18\" >IGVLocalComponent</th>\n",
       "      <td id=\"T_fd7f9_row18_col0\" class=\"data row18 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/MutationReviewer/blob/master/MutationReviewer/AppComponents/IGVLocalComponent.py\">IGVLocalComponent.py</a></td>\n",
       "      <td id=\"T_fd7f9_row18_col1\" class=\"data row18 col1\" >\n",
       "    Callback function to run when the data to review changes\n",
       "    \n",
       "    Parameters\n",
       "    ----------\n",
       "    update_tracks_n_clicks: State\n",
       "        Dash.State of the number of times a button was clicked\n",
       "        \n",
       "    bam_table: State\n",
       "        Dash.State object referencing a state of a dash component containing \n",
       "        a table with the bam files\n",
       "        \n",
       "    bam_table_selected_rows: State, list\n",
       "        Dash.State object referencing a state of a dash component referencing \n",
       "        which rows are selected in a table with the bam files\n",
       "        \n",
       "    gen_data_mut_index_name_func: func\n",
       "        Function used to parse the index to filter the mutation table\n",
       "        \n",
       "    Return\n",
       "    ------\n",
       "    \n",
       "    No updates to the component will be made.\n",
       "    \n",
       "    </td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level2_row19\" class=\"row_heading level2 row19\" >MutationTableComponent</th>\n",
       "      <td id=\"T_fd7f9_row19_col0\" class=\"data row19 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/MutationReviewer/blob/master/MutationReviewer/AppComponents/MutationTableComponent.py\">MutationTableComponent.py</a></td>\n",
       "      <td id=\"T_fd7f9_row19_col1\" class=\"data row19 col1\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level2_row20\" class=\"row_heading level2 row20\" >utils</th>\n",
       "      <td id=\"T_fd7f9_row20_col0\" class=\"data row20 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/MutationReviewer/blob/master/MutationReviewer/AppComponents/utils.py\">utils.py</a></td>\n",
       "      <td id=\"T_fd7f9_row20_col1\" class=\"data row20 col1\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level1_row21\" class=\"row_heading level1 row21\" rowspan=\"2\">Reviewer</th>\n",
       "      <th id=\"T_fd7f9_level2_row21\" class=\"row_heading level2 row21\" >DeprecatedMutationReviewer</th>\n",
       "      <td id=\"T_fd7f9_row21_col0\" class=\"data row21 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/MutationReviewer/blob/master/MutationReviewer/Reviewers/DeprecatedMutationReviewer.py\">DeprecatedMutationReviewer.py</a></td>\n",
       "      <td id=\"T_fd7f9_row21_col1\" class=\"data row21 col1\" >\n",
       "        Parameters\n",
       "        ==========\n",
       "        mutations_df: pd.DataFrame,\n",
       "        bams_df: pd.DataFrame,\n",
       "        mutations_df_sample_col: str,\n",
       "        chrom_col: str,\n",
       "        start_pos_col: str,\n",
       "        bam_df_sample_col: str,\n",
       "        bam_col: str,\n",
       "        bai_col: str,\n",
       "        </td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level2_row22\" class=\"row_heading level2 row22\" >GeneralMutationReviewer</th>\n",
       "      <td id=\"T_fd7f9_row22_col0\" class=\"data row22 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/MutationReviewer/blob/master/MutationReviewer/Reviewers/GeneralMutationReviewer.py\">GeneralMutationReviewer.py</a></td>\n",
       "      <td id=\"T_fd7f9_row22_col1\" class=\"data row22 col1\" >\n",
       "        Parameters\n",
       "        ----------\n",
       "        mutations_df: pd.DataFrame\n",
       "            Table of mutations to review. Each observed mutation has its own row (ie one mutation in each sample). \n",
       "            It must have columns containing information about:\n",
       "                - Chromosome(s)\n",
       "                - Position(s)\n",
       "            \n",
       "        mutation_groupby_cols: list\n",
       "            Additional columns to group mutations in addition to the chromosome and position. \n",
       "            For example, you may group across a participant to view the same locus across all the samples from that participant. \n",
       "            You will need to have a column with the participant value in this table.\n",
       "            \n",
       "        mutations_df_bam_ref_col: str\n",
       "            Column in mutations_df to reference which bam files to view in IGV from bams_df. \n",
       "            If you want to view all samples from the same participant (ie tumor and normal, or serial tumor samples), \n",
       "            the values in \"mutations_df_bam_ref_col\" must correspond to the values in the \"bams_df_ref_col\" column in bams_df.\n",
       "        \n",
       "        chrom_cols: Union[str, list]\n",
       "            Column name(s) in mutations_df referencing the chromosome of the mutation. \n",
       "            You may have multiple loci if you are reviewing a large SV with two breakpoints ([chr1, chr2]).\n",
       "            \n",
       "        pos_cols: Union[str, list]\n",
       "            Column name(s) in mutations_df referencing the position of the mutation. \n",
       "            You may have multiple loci if you are reviewing a large SV with two breakpoints ([chr1, chr2]).\n",
       "            \n",
       "        bams_df: pd.DataFrame\n",
       "            Table with bam paths or urls\n",
       "            \n",
       "        bams_df_ref_col: str\n",
       "            Column in bams_df with values to extract the relevant bams for a given mutation's mutations_df_bam_ref_col value \n",
       "            \n",
       "        bam_cols: Union[str, list]\n",
       "            Column(s) with bams to view\n",
       "            \n",
       "        bai_cols: Union[str, list]\n",
       "            Column(s) with corresponding bai files to view bams. Corresponding bai columns must be in same order as the bam_cols\n",
       "        </td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level0_row23\" class=\"row_heading level0 row23\" rowspan=\"2\">getzlab/PhylogicReviewer</th>\n",
       "      <th id=\"T_fd7f9_level1_row23\" class=\"row_heading level1 row23\" >Repository</th>\n",
       "      <th id=\"T_fd7f9_level2_row23\" class=\"row_heading level2 row23\" >getzlab/PhylogicReviewer</th>\n",
       "      <td id=\"T_fd7f9_row23_col0\" class=\"data row23 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/PhylogicReviewer\">PhylogicReviewer</a></td>\n",
       "      <td id=\"T_fd7f9_row23_col1\" class=\"data row23 col1\" >Interactive app to review Phylogic solutions and data.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_fd7f9_level1_row24\" class=\"row_heading level1 row24\" >Reviewer</th>\n",
       "      <th id=\"T_fd7f9_level2_row24\" class=\"row_heading level2 row24\" >PhylogicReviewer</th>\n",
       "      <td id=\"T_fd7f9_row24_col0\" class=\"data row24 col0\" ><a target=\"_blank\" href=\"https://github.com/getzlab/PhylogicReviewer/blob/master/PhylogicReviewer/Reviewers/PhylogicReviewer.py\">PhylogicReviewer.py</a></td>\n",
       "      <td id=\"T_fd7f9_row24_col1\" class=\"data row24 col1\" >Class to facilitate reviewing Phylogic results in a consistent and efficient manner.\n",
       "\n",
       "    Notes\n",
       "    -----\n",
       "    - Display CCF plot (pull code from Patient Reviewer)\n",
       "    - Display mutations (pull code from Patient Reviewer)\n",
       "    - Display trees (pull code from Patient Reviewer?)\n",
       "        - Work on weighted tree viz\n",
       "    - Display CN plots (pull code from old phylogic review code, with updated library use)**\n",
       "         - Display mutations on CN plots\n",
       "         - Allow filtering using mutations table\n",
       "    - Metrics for coding vs. non_coding, silent vs. non_syn (coding), indels vs. SNPs summarized for each cluster**\n",
       "        - Could output some statistics as well -> highlight sig differences from null\n",
       "    - Integrated (link to run bash script) variant reviewer\n",
       "    - View CCF pmf distributions for individual mutations**\n",
       "\n",
       "    - Annotate:\n",
       "        - Variants (add to variant_review file)\n",
       "        - Cluster annotations (add to cluster blocklist? Only for final analysis though - better to block mutations)\n",
       "            - Can also just annotate with notes\n",
       "        - Notes generally\n",
       "        - Select correct tree (save tree child-parent relationship, but how to associate clusters if you re-run?)\n",
       "    </td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x7fa50c65cfd0>"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "display_catalog_df()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a50385aa-5bae-4d3c-87da-129e7ec2c83f",
   "metadata": {},
   "source": [
    "# Add a Reviewer Suite to Catalog\n",
    "\n",
    "\n",
    "1. Set up repo\n",
    "    Registered repos must satisfy the following:\n",
    "    - Public\n",
    "    - Has a description\n",
    "    - Structured like so (see [AnnoMateTemplate](https://github.com/getzlab/AnnoMateTemplate))\n",
    "        ```\n",
    "        MyReviewers/\n",
    "            MyReviewers/\n",
    "                Reviewers/\n",
    "                    Reviewer1.py\n",
    "                    Reviewer2.py\n",
    "                    ...\n",
    "                AppComponents/\n",
    "                    AppComponent1.py\n",
    "                    AppComponent2.py\n",
    "                    ...\n",
    "        ```\n",
    "   - `setup.py` file\n",
    "\n",
    "1. Fork AnnoMate and go to `example_notebooks/ReviewerCatalog.py`\n",
    "    1. Add the github url of your reviewer repo to the variable `registered_reviewer_repos`\n",
    "    1. Go to `example_notebooks/ReviewerRegistry.ipynb` and run `gen_reviewer_catalog()` (below)\n",
    "    1. Check `catalog.tsv`\n",
    "    \n",
    "1. Commit and submit a pull request"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "aa8177db-03ec-4e28-bed7-a04af99e57b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from ReviewerCatalog import gen_reviewer_catalog, registered_reviewer_repos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "208053be-80ea-4692-b3df-3f54332ce435",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████| 5/5 [00:04<00:00,  1.21it/s]\n"
     ]
    }
   ],
   "source": [
    "gen_reviewer_catalog()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01ae1862-e356-4812-88da-9e55bc2842ca",
   "metadata": {},
   "source": [
    "# Create a reviewer suite\n",
    "\n",
    "See [Developer_AnnoMate_Tutorial.ipynb](https://github.com/getzlab/AnnoMate/blob/master/example_notebooks/Developer_AnnoMate_Tutorial.ipynb) for instructions on how to create a reviewer."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "clean_annomate_env",
   "language": "python",
   "name": "clean_annomate_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
