{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "79f61028",
   "metadata": {},
   "source": [
    "# Reviewer_Tool_dev\n",
    "\n",
    "JupyterReviewer is a package that integrates the manual review processes into Jupyter notebooks and computational analysis workflows."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f60f522",
   "metadata": {},
   "source": [
    "# ReviewData object\n",
    "\n",
    "The `ReviewData` object stores all relevant information regarding the data you need to review. The object is designed to eventually add or edit information for each item (row). Features include:\n",
    "\n",
    "- Organized subtables for data you want to edit, supplementary information to view, and history of changes\n",
    "- Stores subtables automatically\n",
    "- Prevents overwriting\n",
    "- Easy to share or pass review to other users\n",
    "\n",
    "Instantiating a Review Data object requires a dataframe where each row corresponds to the item you want to review (like a mutation or a sample purity). Each row must have some unique index name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8ead7f58",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "prop_id: APP-dropdown-data-state\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import pandas as pd\n",
    "import pathlib\n",
    "import os\n",
    "from IPython.display import display\n",
    "from datetime import datetime, timedelta\n",
    "import time\n",
    "\n",
    "import plotly.express as px\n",
    "from plotly.subplots import make_subplots\n",
    "from jupyter_dash import JupyterDash\n",
    "from dash import dcc\n",
    "from dash import html\n",
    "from dash.dependencies import Input, Output, State\n",
    "from dash.exceptions import PreventUpdate\n",
    "from dash import Dash, dash_table\n",
    "import dash\n",
    "import dash_bootstrap_components as dbc\n",
    "import plotly.express as px"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c5dc5f50",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from datetime import datetime\n",
    "import os\n",
    "import numpy as np\n",
    "import warnings\n",
    "\n",
    "from enum import Enum\n",
    "class AnnotationType(Enum):\n",
    "    TEXT = 'text'\n",
    "    TEXTAREA = 'textarea'\n",
    "    NUMBER = 'number'\n",
    "    CHECKLIST = 'checklist'\n",
    "    RADIOITEM = 'radioitem'\n",
    "\n",
    "class ReviewDataAnnotation:\n",
    "    \n",
    "    def __init__(self, name, \n",
    "                 annot_type: AnnotationType, \n",
    "                 options: []=[], \n",
    "                 validate_input=None,\n",
    "                 default=None\n",
    "                ):\n",
    "        '''\n",
    "        validate_input: a custom function to verify input. Returns a boolean\n",
    "        '''\n",
    "        self.name = name\n",
    "        self.annot_type = annot_type\n",
    "        self.options = options\n",
    "        self.validate_input = validate_input\n",
    "        self.default = default\n",
    "        \n",
    "    def validate(self, x):\n",
    "        if len(self.options) > 0:\n",
    "            for item in np.array([x]).flatten():\n",
    "                if item not in self.options:\n",
    "                    raise ValueError(f'Input {item} is not in the specified options {self.options} for annotation named {self.name}')\n",
    "                \n",
    "        if self.validate_input is not None:\n",
    "            if not self.validate_input(x):\n",
    "                raise ValueError(f'Input {x} is invalid for annotation {self.name}. Check validate_input method')\n",
    "        \n",
    "\n",
    "class ReviewData:\n",
    "    \n",
    "    def __init__(self, \n",
    "                 review_dir: str, # path to directory to save info\n",
    "                 df: pd.DataFrame, # optional if directory above already exists. \n",
    "                 annotate_data: [ReviewDataAnnotation], # dictionary naming column and type of data (text, float, checkbox, radio)\n",
    "                ):\n",
    "        # check df index\n",
    "        \n",
    "        annotate_cols = [ann.name for ann in annotate_data]\n",
    "        self.annotate_data = annotate_data\n",
    "        \n",
    "        self.review_dir = review_dir\n",
    "        self.data_fn = f'{review_dir}/data.tsv'\n",
    "        self.annot_fn = f'{review_dir}/annot.tsv'\n",
    "        self.history_fn = f'{review_dir}/history.tsv'\n",
    "        \n",
    "        if not os.path.isdir(self.review_dir):\n",
    "            os.mkdir(self.review_dir)\n",
    "            self.data = df\n",
    "            self.data.to_csv(self.data_fn, sep='\\t')\n",
    "            self.annot = pd.DataFrame(index=df.index, columns=annotate_cols) # Add more columns. If updating an existing column, will make a new one\n",
    "            self.annot.to_csv(self.annot_fn, sep='\\t')\n",
    "            self.history = pd.DataFrame(columns=annotate_cols + ['index', 'timestamp']) # track all the manual changes, including time stamp\n",
    "            self.history.to_csv(self.history_fn, sep='\\t')\n",
    "        else:\n",
    "            self.data = pd.read_csv(self.data_fn, sep='\\t', index_col=0)\n",
    "            self.annot = pd.read_csv(self.annot_fn, sep='\\t', index_col=0)\n",
    "            self.history = pd.read_csv(self.history_fn, sep='\\t', index_col=0)\n",
    "            \n",
    "        # Add additional annotation columns\n",
    "        new_annot_cols = [c for c in annotate_cols if c not in self.annot.columns]\n",
    "        self.annot[new_annot_cols] = np.nan\n",
    "        \n",
    "        for annot in self.annotate_data:\n",
    "            if annot.annot_type in [AnnotationType.CHECKLIST, AnnotationType.RADIOITEM]:\n",
    "                self.annot[annot_col] = self.annot[annot_col].astype(object)\n",
    "        \n",
    "        # Add additional columns to table\n",
    "        if not df.equals(self.data):\n",
    "            new_data_cols = [c for c in df.columns if c not in self.data.columns]\n",
    "            not_new_data_cols = [c for c in df.columns if c in self.data.columns]\n",
    "            self.data[new_data_cols] = df[new_data_cols]\n",
    "            \n",
    "            if not self.data[not_new_data_cols].equals(df[not_new_data_cols]):\n",
    "                warnings.warn(f'Input data dataframe shares columns with existing data, but are not equal.\\n' + \n",
    "                              f'Only adding columns {new_data_cols} to the ReviewData.data dataframe\\n' + \n",
    "                              f'Remaining columns are not going to be updated.' + \n",
    "                              f'If you intend to change the ReviewData.data attribute, make a new session directory and prefill the annotation data')\n",
    "            \n",
    "    def pre_fill_annot(df: pd.DataFrame):\n",
    "        self.annot.loc[df.index, [c for c in df.columns if c in self.annot.columns]] = df\n",
    "        \n",
    "    def _update(self, data_idx, series):\n",
    "        self.annot.loc[data_idx, list(series.keys())] = list(series.values())\n",
    "        series['timestamp'] = datetime.today()\n",
    "        series['index'] = data_idx\n",
    "        self.history = self.history.append(series, ignore_index=True)\n",
    "        \n",
    "        # write to file\n",
    "        self.data.to_csv(self.data_fn, sep='\\t')\n",
    "        self.annot.to_csv(self.annot_fn, sep='\\t')\n",
    "        self.history.to_csv(self.history_fn, sep='\\t')\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ed979494",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BETA_FLAG_not_enough_drivers</th>\n",
       "      <th>BETA_annot_maf_fn</th>\n",
       "      <th>BETA_clonal_muts</th>\n",
       "      <th>BETA_clonal_muts_genes</th>\n",
       "      <th>BETA_half_purity</th>\n",
       "      <th>BETA_has_beta_solution</th>\n",
       "      <th>BETA_num_clonal_drivers</th>\n",
       "      <th>BETA_ploidy</th>\n",
       "      <th>BETA_purity</th>\n",
       "      <th>BETA_purity_lower</th>\n",
       "      <th>...</th>\n",
       "      <th>manual_purity</th>\n",
       "      <th>manual_purity_lower</th>\n",
       "      <th>manual_purity_upper</th>\n",
       "      <th>manual_ploidy</th>\n",
       "      <th>manual_confidence</th>\n",
       "      <th>manual_flags</th>\n",
       "      <th>last_manual_update</th>\n",
       "      <th>manual_method</th>\n",
       "      <th>MAFLITE</th>\n",
       "      <th>VCF</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>000725_ZS_2668</th>\n",
       "      <td>True</td>\n",
       "      <td>/home/cchu/cgaprojects_ibm_tAML_analysis/data/...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.630</td>\n",
       "      <td>0.570</td>\n",
       "      <td>0.690</td>\n",
       "      <td>2.01</td>\n",
       "      <td>No purity called, unsure</td>\n",
       "      <td>Post_Allo</td>\n",
       "      <td>2022-02-23 21:43:37.104717</td>\n",
       "      <td>Manual_Other</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>005982_GD_1875</th>\n",
       "      <td>False</td>\n",
       "      <td>/home/cchu/cgaprojects_ibm_tAML_analysis/data/...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>['RTEL1:p.A1062T']</td>\n",
       "      <td>0.488</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.976</td>\n",
       "      <td>0.860</td>\n",
       "      <td>...</td>\n",
       "      <td>0.910</td>\n",
       "      <td>0.860</td>\n",
       "      <td>0.960</td>\n",
       "      <td>1.95</td>\n",
       "      <td>Confident</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-02-23 21:44:00.961518</td>\n",
       "      <td>Keep_auto_call</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>012413_AT_1634</th>\n",
       "      <td>False</td>\n",
       "      <td>/home/cchu/cgaprojects_ibm_tAML_analysis/data/...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>['BRCA1:p.V772A']</td>\n",
       "      <td>0.512</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.024</td>\n",
       "      <td>0.800</td>\n",
       "      <td>...</td>\n",
       "      <td>1.024</td>\n",
       "      <td>0.800</td>\n",
       "      <td>1.244</td>\n",
       "      <td>2.00</td>\n",
       "      <td>Purity called, unsure</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-02-23 21:44:51.038330</td>\n",
       "      <td>Manual_BETA</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>016198_VX_1736</th>\n",
       "      <td>False</td>\n",
       "      <td>/home/cchu/cgaprojects_ibm_tAML_analysis/data/...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>['TERT:p.R756H']</td>\n",
       "      <td>0.430</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.860</td>\n",
       "      <td>0.760</td>\n",
       "      <td>...</td>\n",
       "      <td>0.860</td>\n",
       "      <td>0.760</td>\n",
       "      <td>0.964</td>\n",
       "      <td>2.00</td>\n",
       "      <td>Purity called, unsure</td>\n",
       "      <td>No CNA</td>\n",
       "      <td>2022-02-23 21:51:14.522862</td>\n",
       "      <td>Manual_BETA</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>022613_PU_3426</th>\n",
       "      <td>True</td>\n",
       "      <td>/home/cchu/cgaprojects_ibm_tAML_analysis/data/...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.460</td>\n",
       "      <td>0.390</td>\n",
       "      <td>0.530</td>\n",
       "      <td>1.88</td>\n",
       "      <td>Confident</td>\n",
       "      <td>No_AML_drivers</td>\n",
       "      <td>2022-02-23 21:53:02.173752</td>\n",
       "      <td>Keep_auto_call</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PQ9867BM</th>\n",
       "      <td>False</td>\n",
       "      <td>/home/cchu/cgaprojects_ibm_tAML_analysis/data/...</td>\n",
       "      <td>[0, 1]</td>\n",
       "      <td>['TP53:p.W53*', 'ZNF318:p.R1936S']</td>\n",
       "      <td>0.400</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.800</td>\n",
       "      <td>0.688</td>\n",
       "      <td>...</td>\n",
       "      <td>0.864</td>\n",
       "      <td>0.708</td>\n",
       "      <td>1.024</td>\n",
       "      <td>2.00</td>\n",
       "      <td>Confident</td>\n",
       "      <td>Used DFCI flags to change Beta solution</td>\n",
       "      <td>2022-03-08 21:01:12.668613</td>\n",
       "      <td>Manual_BETA</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SA04142016</th>\n",
       "      <td>True</td>\n",
       "      <td>/home/cchu/cgaprojects_ibm_tAML_analysis/data/...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.920</td>\n",
       "      <td>0.870</td>\n",
       "      <td>0.970</td>\n",
       "      <td>1.83</td>\n",
       "      <td>Confident</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-03-08 21:01:38.015167</td>\n",
       "      <td>Keep_auto_call</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SM120519BM-H</th>\n",
       "      <td>True</td>\n",
       "      <td>/home/cchu/cgaprojects_ibm_tAML_analysis/data/...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>No purity called, unsure</td>\n",
       "      <td>No CNA,No AML drivers</td>\n",
       "      <td>2022-03-08 21:02:25.093167</td>\n",
       "      <td>Manual_Other</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WD10052017BM</th>\n",
       "      <td>False</td>\n",
       "      <td>/home/cchu/cgaprojects_ibm_tAML_analysis/data/...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>['CTC1:p.R731W']</td>\n",
       "      <td>0.700</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.400</td>\n",
       "      <td>0.732</td>\n",
       "      <td>...</td>\n",
       "      <td>0.528</td>\n",
       "      <td>0.464</td>\n",
       "      <td>0.596</td>\n",
       "      <td>2.00</td>\n",
       "      <td>Purity called, unsure</td>\n",
       "      <td>No CNA</td>\n",
       "      <td>2022-03-08 21:04:18.622309</td>\n",
       "      <td>Manual_BETA</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZF022118BM</th>\n",
       "      <td>False</td>\n",
       "      <td>/home/cchu/cgaprojects_ibm_tAML_analysis/data/...</td>\n",
       "      <td>[0, 1]</td>\n",
       "      <td>['RYBP:p.P92S', 'SAMD9:p.R1040H']</td>\n",
       "      <td>0.490</td>\n",
       "      <td>True</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.980</td>\n",
       "      <td>0.916</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.950</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.97</td>\n",
       "      <td>Confident</td>\n",
       "      <td>Used DFCI flags to change Beta solution</td>\n",
       "      <td>2022-03-08 21:07:20.562203</td>\n",
       "      <td>Keep_auto_call</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "      <td>gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>642 rows Ã— 446 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                BETA_FLAG_not_enough_drivers  \\\n",
       "sample_id                                      \n",
       "000725_ZS_2668                          True   \n",
       "005982_GD_1875                         False   \n",
       "012413_AT_1634                         False   \n",
       "016198_VX_1736                         False   \n",
       "022613_PU_3426                          True   \n",
       "...                                      ...   \n",
       "PQ9867BM                               False   \n",
       "SA04142016                              True   \n",
       "SM120519BM-H                            True   \n",
       "WD10052017BM                           False   \n",
       "ZF022118BM                             False   \n",
       "\n",
       "                                                BETA_annot_maf_fn  \\\n",
       "sample_id                                                           \n",
       "000725_ZS_2668  /home/cchu/cgaprojects_ibm_tAML_analysis/data/...   \n",
       "005982_GD_1875  /home/cchu/cgaprojects_ibm_tAML_analysis/data/...   \n",
       "012413_AT_1634  /home/cchu/cgaprojects_ibm_tAML_analysis/data/...   \n",
       "016198_VX_1736  /home/cchu/cgaprojects_ibm_tAML_analysis/data/...   \n",
       "022613_PU_3426  /home/cchu/cgaprojects_ibm_tAML_analysis/data/...   \n",
       "...                                                           ...   \n",
       "PQ9867BM        /home/cchu/cgaprojects_ibm_tAML_analysis/data/...   \n",
       "SA04142016      /home/cchu/cgaprojects_ibm_tAML_analysis/data/...   \n",
       "SM120519BM-H    /home/cchu/cgaprojects_ibm_tAML_analysis/data/...   \n",
       "WD10052017BM    /home/cchu/cgaprojects_ibm_tAML_analysis/data/...   \n",
       "ZF022118BM      /home/cchu/cgaprojects_ibm_tAML_analysis/data/...   \n",
       "\n",
       "               BETA_clonal_muts              BETA_clonal_muts_genes  \\\n",
       "sample_id                                                             \n",
       "000725_ZS_2668              NaN                                 NaN   \n",
       "005982_GD_1875              [0]                  ['RTEL1:p.A1062T']   \n",
       "012413_AT_1634              [0]                   ['BRCA1:p.V772A']   \n",
       "016198_VX_1736              [0]                    ['TERT:p.R756H']   \n",
       "022613_PU_3426              NaN                                 NaN   \n",
       "...                         ...                                 ...   \n",
       "PQ9867BM                 [0, 1]  ['TP53:p.W53*', 'ZNF318:p.R1936S']   \n",
       "SA04142016                  NaN                                 NaN   \n",
       "SM120519BM-H                NaN                                 NaN   \n",
       "WD10052017BM                [0]                    ['CTC1:p.R731W']   \n",
       "ZF022118BM               [0, 1]   ['RYBP:p.P92S', 'SAMD9:p.R1040H']   \n",
       "\n",
       "                BETA_half_purity  BETA_has_beta_solution  \\\n",
       "sample_id                                                  \n",
       "000725_ZS_2668               NaN                   False   \n",
       "005982_GD_1875             0.488                    True   \n",
       "012413_AT_1634             0.512                    True   \n",
       "016198_VX_1736             0.430                    True   \n",
       "022613_PU_3426               NaN                   False   \n",
       "...                          ...                     ...   \n",
       "PQ9867BM                   0.400                    True   \n",
       "SA04142016                   NaN                   False   \n",
       "SM120519BM-H                 NaN                   False   \n",
       "WD10052017BM               0.700                    True   \n",
       "ZF022118BM                 0.490                    True   \n",
       "\n",
       "                BETA_num_clonal_drivers  BETA_ploidy  BETA_purity  \\\n",
       "sample_id                                                           \n",
       "000725_ZS_2668                      NaN          2.0        0.000   \n",
       "005982_GD_1875                      1.0          2.0        0.976   \n",
       "012413_AT_1634                      2.0          2.0        1.024   \n",
       "016198_VX_1736                      1.0          2.0        0.860   \n",
       "022613_PU_3426                      NaN          2.0        0.000   \n",
       "...                                 ...          ...          ...   \n",
       "PQ9867BM                            2.0          2.0        0.800   \n",
       "SA04142016                          NaN          2.0        0.000   \n",
       "SM120519BM-H                        NaN          2.0        0.000   \n",
       "WD10052017BM                        2.0          2.0        1.400   \n",
       "ZF022118BM                          5.0          2.0        0.980   \n",
       "\n",
       "                BETA_purity_lower  ...  manual_purity manual_purity_lower  \\\n",
       "sample_id                          ...                                      \n",
       "000725_ZS_2668              0.000  ...          0.630               0.570   \n",
       "005982_GD_1875              0.860  ...          0.910               0.860   \n",
       "012413_AT_1634              0.800  ...          1.024               0.800   \n",
       "016198_VX_1736              0.760  ...          0.860               0.760   \n",
       "022613_PU_3426              0.000  ...          0.460               0.390   \n",
       "...                           ...  ...            ...                 ...   \n",
       "PQ9867BM                    0.688  ...          0.864               0.708   \n",
       "SA04142016                  0.000  ...          0.920               0.870   \n",
       "SM120519BM-H                0.000  ...          0.000               0.000   \n",
       "WD10052017BM                0.732  ...          0.528               0.464   \n",
       "ZF022118BM                  0.916  ...          1.000               0.950   \n",
       "\n",
       "               manual_purity_upper manual_ploidy         manual_confidence  \\\n",
       "sample_id                                                                    \n",
       "000725_ZS_2668               0.690          2.01  No purity called, unsure   \n",
       "005982_GD_1875               0.960          1.95                 Confident   \n",
       "012413_AT_1634               1.244          2.00     Purity called, unsure   \n",
       "016198_VX_1736               0.964          2.00     Purity called, unsure   \n",
       "022613_PU_3426               0.530          1.88                 Confident   \n",
       "...                            ...           ...                       ...   \n",
       "PQ9867BM                     1.024          2.00                 Confident   \n",
       "SA04142016                   0.970          1.83                 Confident   \n",
       "SM120519BM-H                 0.000          0.00  No purity called, unsure   \n",
       "WD10052017BM                 0.596          2.00     Purity called, unsure   \n",
       "ZF022118BM                   1.000          1.97                 Confident   \n",
       "\n",
       "                                           manual_flags  \\\n",
       "sample_id                                                 \n",
       "000725_ZS_2668                                Post_Allo   \n",
       "005982_GD_1875                                      NaN   \n",
       "012413_AT_1634                                      NaN   \n",
       "016198_VX_1736                                   No CNA   \n",
       "022613_PU_3426                           No_AML_drivers   \n",
       "...                                                 ...   \n",
       "PQ9867BM        Used DFCI flags to change Beta solution   \n",
       "SA04142016                                          NaN   \n",
       "SM120519BM-H                      No CNA,No AML drivers   \n",
       "WD10052017BM                                     No CNA   \n",
       "ZF022118BM      Used DFCI flags to change Beta solution   \n",
       "\n",
       "                        last_manual_update   manual_method  \\\n",
       "sample_id                                                    \n",
       "000725_ZS_2668  2022-02-23 21:43:37.104717    Manual_Other   \n",
       "005982_GD_1875  2022-02-23 21:44:00.961518  Keep_auto_call   \n",
       "012413_AT_1634  2022-02-23 21:44:51.038330     Manual_BETA   \n",
       "016198_VX_1736  2022-02-23 21:51:14.522862     Manual_BETA   \n",
       "022613_PU_3426  2022-02-23 21:53:02.173752  Keep_auto_call   \n",
       "...                                    ...             ...   \n",
       "PQ9867BM        2022-03-08 21:01:12.668613     Manual_BETA   \n",
       "SA04142016      2022-03-08 21:01:38.015167  Keep_auto_call   \n",
       "SM120519BM-H    2022-03-08 21:02:25.093167    Manual_Other   \n",
       "WD10052017BM    2022-03-08 21:04:18.622309     Manual_BETA   \n",
       "ZF022118BM      2022-03-08 21:07:20.562203  Keep_auto_call   \n",
       "\n",
       "                                                          MAFLITE  \\\n",
       "sample_id                                                           \n",
       "000725_ZS_2668  gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...   \n",
       "005982_GD_1875  gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...   \n",
       "012413_AT_1634  gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...   \n",
       "016198_VX_1736  gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...   \n",
       "022613_PU_3426  gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...   \n",
       "...                                                           ...   \n",
       "PQ9867BM        gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...   \n",
       "SA04142016      gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...   \n",
       "SM120519BM-H    gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...   \n",
       "WD10052017BM    gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...   \n",
       "ZF022118BM      gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...   \n",
       "\n",
       "                                                              VCF  \n",
       "sample_id                                                          \n",
       "000725_ZS_2668  gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...  \n",
       "005982_GD_1875  gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...  \n",
       "012413_AT_1634  gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...  \n",
       "016198_VX_1736  gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...  \n",
       "022613_PU_3426  gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...  \n",
       "...                                                           ...  \n",
       "PQ9867BM        gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...  \n",
       "SA04142016      gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...  \n",
       "SM120519BM-H    gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...  \n",
       "WD10052017BM    gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...  \n",
       "ZF022118BM      gs://fc-fed5ee4d-4de5-429a-b88e-681cde1f0558/a...  \n",
       "\n",
       "[642 rows x 446 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bucket_0c1_cchu_manual_purity_review_session_dir = 'gs://taml_vm_analysis/data/Full-Analysis/1_Full-Analysis-2022-02-22_pran3/0c1_Manual_Purity_Review_cchu'\n",
    "cchu_purities_df = pd.read_csv(f'{bucket_0c1_cchu_manual_purity_review_session_dir}/manual_purity_review_table.tsv', sep='\\t', index_col=0)\n",
    "cchu_purities_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ef9e93e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>purity</th>\n",
       "      <th>class</th>\n",
       "      <th>rating</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>000725_ZS_2668</th>\n",
       "      <td>2.0</td>\n",
       "      <td>Option 3</td>\n",
       "      <td>2.0</td>\n",
       "      <td>asdf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>005982_GD_1875</th>\n",
       "      <td>0.0</td>\n",
       "      <td>Option 0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>012413_AT_1634</th>\n",
       "      <td>0.4</td>\n",
       "      <td>Option 2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>asdfsadf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>016198_VX_1736</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>022613_PU_3426</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                purity     class  rating description\n",
       "sample_id                                           \n",
       "000725_ZS_2668     2.0  Option 3     2.0        asdf\n",
       "005982_GD_1875     0.0  Option 0     0.0           0\n",
       "012413_AT_1634     0.4  Option 2    10.0    asdfsadf\n",
       "016198_VX_1736     NaN       NaN     NaN         NaN\n",
       "022613_PU_3426     NaN       NaN     NaN         NaN"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_rd_dir = '/home/cchu/cgaprojects_ibm_tAML_analysis/data/test_getzlab-JupyterReviewer/Reviewer_Tutorial'\n",
    "test_rd = ReviewData(review_dir=test_rd_dir,\n",
    "                     df = cchu_purities_df, # optional if directory above already exists. \n",
    "                     annotate_data = [ReviewDataAnnotation('purity', 'number', validate_input=lambda x: x < 0.5),\n",
    "                                      ReviewDataAnnotation('rating', 'number', options=range(10)),\n",
    "                                      ReviewDataAnnotation('description', 'text'),\n",
    "                                      ReviewDataAnnotation('class', 'radioitem', options=[f'Option {n}' for n in range(4)]),])\n",
    "                     \n",
    "#                      {'purity': 'number', \n",
    "#                                       'class': 'text', \n",
    "#                                       'rating': 'number', \n",
    "#                                       'description': 'text', \n",
    "#                                       'another_annot_col': 'checklist'})\n",
    "test_rd.annot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "767f6fa2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "bcef14d8",
   "metadata": {},
   "source": [
    "# Simple widgets notebook reviewer\n",
    "\n",
    "You can use ipython widgets to get interactivity "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fa95127",
   "metadata": {},
   "source": [
    "# ReviewDataApp object\n",
    "\n",
    "Use the functionality of ploty Dash to create advanced dashboards for visualizing and interacting with your data. This is made to wrap around any ReviewData object, so it is easy to edit and change as needed without undoing the underlying annotations in the ReviewData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "id": "fa21ce4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "class AppComponent:\n",
    "    \n",
    "    def __init__(self, name, \n",
    "                 components, \n",
    "                 callback=None, \n",
    "                 callback_output=[], \n",
    "                 callback_input=[],\n",
    "                 callback_state=[],\n",
    "                ):\n",
    "        self.name = name\n",
    "        self.component = html.Div(components)\n",
    "        self.callback = callback\n",
    "        self.callback_output = callback_output\n",
    "        self.callback_input = callback_input\n",
    "        self.callback_state = callback_state\n",
    "        \n",
    "        # TODO: option to update anotations\n",
    "        # TODO: reset function (switching samples) and a page function\n",
    "    \n",
    "class TestApp:\n",
    "    def __init__(self, review_data: ReviewData, host='0.0.0.0', port=8051):\n",
    "        self.prop = None\n",
    "        self.autofill_buttons = []\n",
    "        self.autofill_input_dict = {} #{buttonid: {annot_col: Input(compoennt value)}}\n",
    "        self.more_components = []  # TODO: set custom layout?\n",
    "        self.review_data = review_data\n",
    "        self.host = host\n",
    "        self.port = port\n",
    "        \n",
    "        # check component ids are not duplicated\n",
    "        \n",
    "    def run_app(self, mode, host='0.0.0.0', port=8050):\n",
    "        app = JupyterDash(__name__, external_stylesheets=[dbc.themes.BOOTSTRAP])\n",
    "        app.layout = self.gen_layout()\n",
    "\n",
    "        @app.callback(output=dict(history_table=Output(f'APP-history-table', 'children'),\n",
    "                                  annot_panel=self.annotation_panel_component.callback_output,\n",
    "                                  more_component_outputs={c.name: c.callback_output for c in self.more_components}\n",
    "                             ), \n",
    "                      inputs=dict(dropdown_value=Input('APP-dropdown-data-state', 'value'), \n",
    "                                  autofill_buttons=[Input(b.id, 'n_clicks') for b in self.autofill_buttons],\n",
    "                                  autofill_inputs=self.autofill_input_dict,\n",
    "                                  submit_annot_button=Input('APP-submit-button-state', 'n_clicks'),\n",
    "                                  annot_input_state=self.annotation_panel_component.callback_state, #{annot.name: State(f\"APP-{annot.name}-{annot.annot_type}-input-state\", \"value\") for annot in self.review_data.annotate_data},\n",
    "                                  more_component_inputs={c.name: c.callback_input for c in self.more_components}\n",
    "                                 )\n",
    "                     ) # TODO: add back more components\n",
    "        def component_callback(dropdown_value, \n",
    "                               autofill_buttons,\n",
    "                               autofill_inputs,\n",
    "                               submit_annot_button, \n",
    "                               annot_input_state, \n",
    "                               more_component_inputs):\n",
    "            \n",
    "            ctx = dash.callback_context\n",
    "            if not ctx.triggered:\n",
    "                raise PreventUpdate\n",
    "            else:\n",
    "                prop_id = ctx.triggered[0]['prop_id'].split('.')[0]\n",
    "            \n",
    "            output_dict = {'history_table': dash.no_update, \n",
    "                           'annot_panel': {annot_col: dash.no_update for annot_col in self.review_data.annot.columns}, \n",
    "                           'more_component_outputs': {c.name: [dash.no_update for i in range(len(c.callback_output))] for c in self.more_components}}\n",
    "            \n",
    "            if prop_id == 'APP-dropdown-data-state':\n",
    "\n",
    "                for i in range(len(self.more_components)):\n",
    "                    component = self.more_components[i]\n",
    "                    # reset vs row dependent\n",
    "                    component_output = component.callback(self.review_data.data.loc[dropdown_value], *more_component_inputs[component.name])\n",
    "                    output_dict['more_component_outputs'][component.name] = component_output # force this? specify names in the callback outputs?\n",
    "                    \n",
    "                output_dict['history_table'] = dbc.Table.from_dataframe(self.review_data.history.loc[self.review_data.history['index'] == dropdown_value])\n",
    "                output_dict['annot_panel'] = {annot_col: '' for annot_col in self.review_data.annot.columns} # TODO set defaults?\n",
    "                            \n",
    "            elif (prop_id == 'APP-submit-button-state') & (submit_annot_button > 0):\n",
    "                self.review_data._update(dropdown_value, annot_input_state)\n",
    "                output_dict['history_table'] = dbc.Table.from_dataframe(self.review_data.history.loc[self.review_data.history['index'] == dropdown_value])\n",
    "            elif 'APP-autofill-' in prop_id:\n",
    "                component_name = prop_id.split('APP-autofill-')[-1]\n",
    "                for autofill_annot_col, value in autofill_inputs[prop_id].items():\n",
    "                    output_dict['annot_panel'][autofill_annot_col] = value\n",
    "            else:\n",
    "                # identify component that changed and which outputs are changed\n",
    "                for i in range(len(self.more_components)):\n",
    "                    component = self.more_components[i]\n",
    "                    if sum([c.component_id == prop_id for c in self.more_components[i].callback_input]) > 0:\n",
    "                        component_output = component.callback(self.review_data.data.loc[dropdown_value], *more_component_inputs[component.name])\n",
    "                        output_dict['more_component_outputs'][component.name] = component_output # force having output as array specify names in the callback outputs?\n",
    "                pass\n",
    "            return output_dict\n",
    "        \n",
    "        app.run_server(mode=mode, host=host, port=port, debug=True) \n",
    "        \n",
    "        \n",
    "    def gen_annotation_panel_component(self):\n",
    "        annotation_data = self.review_data.annotate_data\n",
    "        \n",
    "        \n",
    "        submit_annot_button = html.Button(id='APP-submit-button-state', n_clicks=0, children='Submit')\n",
    "        \n",
    "        # history panel\n",
    "        \n",
    "        def annotation_input(annot: ReviewDataAnnotation):\n",
    "            \n",
    "            input_component_id = f\"APP-{annot.name}-{annot.annot_type}-input-state\"\n",
    "            \n",
    "            if annot.annot_type == AnnotationType.TEXTAREA.value:\n",
    "                input_component = dbc.Textarea(size=\"lg\", \n",
    "                                               id=input_component_id,\n",
    "                                               value=annot.default,\n",
    "                                              ), \n",
    "            elif annot.annot_type ==  AnnotationType.TEXT.value:\n",
    "                input_component = dbc.Input(type=\"text\", \n",
    "                                    id=input_component_id, \n",
    "                                    placeholder=f\"Enter {annot.name}\",\n",
    "                                    value=annot.default,\n",
    "                                   )\n",
    "            elif annot.annot_type == AnnotationType.NUMBER.value:\n",
    "                input_component = dbc.Input(type=\"number\", \n",
    "                                    id=input_component_id, \n",
    "                                    placeholder=f\"Enter {annot.name}\",\n",
    "                                    value=annot.default,\n",
    "                                   )\n",
    "            elif annot.annot_type == AnnotationType.CHECKLIST.value:\n",
    "                flags = np.arange(0, 1, 0.2)\n",
    "                input_component = dbc.Checklist(options=[{\"label\": f, \"value\": f} for f in annot.options],\n",
    "                                                id=input_component_id, \n",
    "                                                value=annot.default),\n",
    "            elif annot.annot_type == AnnotationType.RADIOITEM.value:\n",
    "                # TODO: how to add in options\n",
    "                input_component = dbc.RadioItems(\n",
    "                                                options=[{\"label\": f, \"value\": f} for f in annot.options],\n",
    "                                                value=annot.default,\n",
    "                                                id=input_component_id,\n",
    "                                            ),\n",
    "            else:\n",
    "                raise ValueError(f'Invalid annotation type \"{annot.annot_type}\"')\n",
    "                \n",
    "            return dbc.Row([dbc.Label(annot.name, html_for=input_component_id, width=2), dbc.Col(input_component)])\n",
    "        \n",
    "        panel_components = self.autofill_buttons + [annotation_input(annot) for annot in self.review_data.annotate_data] + [submit_annot_button]\n",
    "        panel_inputs = [Input('APP-submit-button-state', 'nclicks')]\n",
    "        return AppComponent(name='APP-Panel',\n",
    "                           components=panel_components, \n",
    "                           callback_output={annot.name: Output(f\"APP-{annot.name}-{annot.annot_type}-input-state\", \"value\") for annot in self.review_data.annotate_data},\n",
    "                           callback_input=panel_inputs,\n",
    "                           callback_state={annot.name: State(f\"APP-{annot.name}-{annot.annot_type}-input-state\", \"value\") for annot in self.review_data.annotate_data}\n",
    "                          )\n",
    "        \n",
    "    def gen_layout(self):\n",
    "        \n",
    "        dropdown = html.Div(dcc.Dropdown(options=self.review_data.data.index, \n",
    "                                         value=self.review_data.data.index[0], \n",
    "                                         id='APP-dropdown-data-state'))\n",
    "        \n",
    "        self.dropdown_component = AppComponent(name='APP-dropdown-component',\n",
    "                                               components=[dropdown])\n",
    "        \n",
    "        history_table = html.Div([dbc.Table.from_dataframe(pd.DataFrame(columns=self.review_data.history.columns))], id='APP-history-table')\n",
    "        self.history_component = AppComponent(name='APP-history-component',\n",
    "                                               components=[history_table])\n",
    "        \n",
    "        self.annotation_panel_component = self.gen_annotation_panel_component()\n",
    "        \n",
    "        layout = html.Div([dbc.Row(self.dropdown_component.component, justify='end'),\n",
    "                           dbc.Row([dbc.Col(self.annotation_panel_component.component),\n",
    "                                    dbc.Col(self.history_component.component)\n",
    "                                   ]),\n",
    "                           dbc.Row([dbc.Row(c.component) for c in self.more_components])\n",
    "                          ])\n",
    "\n",
    "        return layout\n",
    "    \n",
    "    def add_table_from_path(self, table_name, component_name, col, table_cols):\n",
    "        \n",
    "        table = html.Div(dbc.Table.from_dataframe(pd.read_csv(self.review_data.data.iloc[0][col], sep='\\t')[table_cols]), \n",
    "                                   id=component_name)\n",
    "        table_component = AppComponent(component_name, [html.H1(table_name), table], \n",
    "                                      lambda r: [dbc.Table.from_dataframe(pd.read_csv(r[col], sep='\\t')[table_cols])],\n",
    "                                      callback_output=[Output(component_name, 'children')], \n",
    "                                     )\n",
    "        self.more_components.append(table_component)\n",
    "        \n",
    "    def add_custom_component(self, \n",
    "                             component_name, \n",
    "                             component_layout,\n",
    "                             func, \n",
    "                             callback_output, \n",
    "                             callback_input=[], \n",
    "                             add_autofill=False,\n",
    "                             autofill_dict={}, # annot_col: component output id\n",
    "                             **kwargs):\n",
    "        \n",
    "        if add_autofill:\n",
    "            autofill_button_component = html.Button(f'Use {component_name} solution', id=f'APP-autofill-{component_name}', n_clicks=0)\n",
    "            self.autofill_buttons += [autofill_button_component]\n",
    "            self.autofill_input_dict[autofill_button_component.id] = autofill_dict\n",
    "            \n",
    "        component = AppComponent(component_name, component_layout, \n",
    "                                  lambda *args: func(*args, **kwargs),\n",
    "                                  callback_output=callback_output, \n",
    "                                  callback_input=callback_input\n",
    "                                 )\n",
    "        \n",
    "        self.more_components.append(component)\n",
    "        \n",
    "        \n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebc57824",
   "metadata": {},
   "source": [
    "# Example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "id": "685d348b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import beta, kruskal\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "tumor_f_bin_width = 1.0/500.0\n",
    "tumor_f_bins = np.arange(0, 1, tumor_f_bin_width)\n",
    "pval_threshold = 1.1E-4\n",
    "def plot_beta(maf_df, data_id):\n",
    "    \n",
    "    if maf_df.empty:\n",
    "        raise ValueError(\"There are no mutations in the maf dataframe.\")\n",
    "\n",
    "    for idx, r in maf_df.iterrows():\n",
    "        pdf = beta.pdf(tumor_f_bins, r['t_alt_count'] + 1, r['t_ref_count'] + 1)\n",
    "        maf_df.loc[idx, tumor_f_bins] = pdf / (sum(pdf) * tumor_f_bin_width)\n",
    "\n",
    "    sum_pdf = maf_df[tumor_f_bins].sum(axis=0)\n",
    "    sum_pdf = sum_pdf / (sum_pdf.sum() * tumor_f_bin_width)\n",
    "    if 'tumor_f' not in maf_df.columns:\n",
    "        maf_df['tumor_f'] = maf_df['t_alt_count'].astype(float) / (maf_df['t_alt_count'] + maf_df['t_ref_count'])\n",
    "    maf_df = maf_df.sort_values(by='tumor_f',\n",
    "                                ascending=False).reset_index()\n",
    "\n",
    "    clonal_muts = [maf_df.index[0]]  # Get the first one\n",
    "    for j in np.arange(maf_df.shape[0], 1, -1):\n",
    "        h_stat, pval = kruskal(*maf_df.iloc[:j].apply(lambda x: np.concatenate((np.ones(x['t_alt_count']),\n",
    "                                                                                np.zeros(x['t_ref_count']))),\n",
    "                                                      axis=1).tolist())\n",
    "        if pval > pval_threshold:\n",
    "            clonal_muts = maf_df.index[:j].tolist()\n",
    "            break\n",
    "\n",
    "    subclonal_muts = maf_df.index[clonal_muts[-1] + 1:].tolist() if clonal_muts[-1] < maf_df.shape[0] else []\n",
    "\n",
    "    clonal_prod_pdf = maf_df.loc[clonal_muts, tumor_f_bins].product(axis=0)\n",
    "    clonal_prod_pdf = clonal_prod_pdf / (clonal_prod_pdf.sum() * tumor_f_bin_width)\n",
    "    half_purity = clonal_prod_pdf.argmax()\n",
    "    purity = clonal_prod_pdf.index[half_purity] * 2\n",
    "\n",
    "    log_clonal_prod_pdf = np.log10(clonal_prod_pdf)\n",
    "    log_clonal_prod_pdf = log_clonal_prod_pdf - np.max(log_clonal_prod_pdf)\n",
    "    cis = log_clonal_prod_pdf[log_clonal_prod_pdf >= -1].index.tolist()\n",
    "    purity_lower_ci = cis[0] * 2\n",
    "    purity_upper_ci = cis[-1] * 2\n",
    "    \n",
    "    # plotly plot\n",
    "    # Step 1: make the figure\n",
    "    maf_df['clonal_status'] = maf_df.index.map(lambda x: 'clonal' if x in clonal_muts else 'subclonal')\n",
    "    maf_df['Mut_Label'] = maf_df['Hugo_Symbol'] + ':' + maf_df['Start_position'].astype(str) + ':' + maf_df['Protein_Change'].astype(str) + ':' + maf_df['Variant_Classification'].astype(str)\n",
    "    to_plot_maf_df = maf_df.set_index('Mut_Label')[list(tumor_f_bins)].stack().reset_index()\n",
    "    to_plot_maf_df['clonal_status'] = to_plot_maf_df['Mut_Label'].map(maf_df[['Mut_Label', 'clonal_status']].set_index('Mut_Label')['clonal_status'])\n",
    "    to_plot_maf_df['pdf_log10'] = np.log10(to_plot_maf_df[0])\n",
    "    fig = px.line(to_plot_maf_df, x='level_1', y='pdf_log10', color='clonal_status', \n",
    "                  hover_data=['Mut_Label'], title=f'{data_id}: purity = {round(purity, 2)} [{round(purity_lower_ci, 2)} - {round(purity_upper_ci, 2)}]')\n",
    "    fig.add_trace(go.Scatter(x=tumor_f_bins, y=np.log10(clonal_prod_pdf),\n",
    "                    mode='lines',\n",
    "                    name='clonal product pdf'))\n",
    "    fig.add_trace(go.Scatter(x=tumor_f_bins, y=np.log10(sum_pdf),\n",
    "                    mode='lines',\n",
    "                    name='all mutations sum pdf'))\n",
    "    \n",
    "    fig.add_vrect(x0=cis[0], x1=cis[-1], line_width=0, fillcolor=\"red\", opacity=0.2)\n",
    "    fig.add_vline(x=clonal_prod_pdf.index[half_purity], name='Half purity')\n",
    "    \n",
    "    ylim_min=10 ** (-4)\n",
    "    ylim_max=10 ** 2\n",
    "    fig.update_yaxes(range=[np.log10(ylim_min), np.log10(ylim_max)])\n",
    "\n",
    "    return fig, purity, purity_lower_ci, purity_upper_ci\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "id": "d6c3e95e",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "import functools\n",
    "import time\n",
    "\n",
    "test_app = TestApp(test_rd)\n",
    "\n",
    "# add components\n",
    "def gen_data_summary_table(r, cols):\n",
    "    return [[html.H1(f'{r.name} Data Summary'), dbc.Table.from_dataframe(r[cols].to_frame().reset_index())]]\n",
    "\n",
    "test_app.add_custom_component('sample-info-component', \n",
    "                              html.Div(children=[html.H1('Data Summary'), \n",
    "                                                 dbc.Table.from_dataframe(df=pd.DataFrame())],\n",
    "                                       id='sample-info-component'\n",
    "                                      ), \n",
    "                              callback_output=[Output('sample-info-component', 'children')],\n",
    "                              func=gen_data_summary_table, \n",
    "                              cols=['BETA_ploidy',\n",
    "                                     'BETA_purity',\n",
    "                                     'BETA_purity_lower',\n",
    "                                     'BETA_purity_upper'])\n",
    "\n",
    "test_app.add_table_from_path('DFCI MAF file', 'maf-component-id', 'DFCI_local_sample_dfci_maf_fn', ['Hugo_Symbol', 'Chromosome', 't_alt_count', 't_ref_count', 'Tumor_Sample_Barcode'])\n",
    "\n",
    "beta_table_cols = ['CHIP_mut_status', \n",
    "                  'aSCNA', \n",
    "                  'Hugo_Symbol', \n",
    "                  'Chromosome', \n",
    "                  'Start_position', \n",
    "                  'Variant_Classification', \n",
    "                  'Protein_Change', \n",
    "                  't_alt_count', \n",
    "                  't_ref_count', \n",
    "                  'total_count', \n",
    "                  'tumor_f', \n",
    "                  'gnomADg_AF']\n",
    "blank_beta_df = pd.DataFrame(columns=beta_table_cols)\n",
    "blank_beta_df.loc[0, beta_table_cols] = 'Test'\n",
    "\n",
    "@functools.lru_cache(maxsize=32) # faster to reload\n",
    "def read_maf(fn):\n",
    "    return pd.read_csv(fn, sep='\\t')\n",
    "\n",
    "def beta_graph_callback(r, \n",
    "                        reload_beta_graph_button, \n",
    "                        selected_rows, \n",
    "                        beta_table_fn_col, \n",
    "                        beta_table_display_col):\n",
    "    maf_df = read_maf(r[beta_table_fn_col])\n",
    "    fig, purity, purity_lower_ci, purity_upper_ci = plot_beta(maf_df.loc[selected_rows], r.name)\n",
    "    return [fig, maf_df[beta_table_display_col].to_dict('records'), selected_rows, purity, 2]\n",
    "    \n",
    "test_app.add_custom_component('beta-graph', \n",
    "                              html.Div([html.H1(\"Beta MAF\"), \n",
    "                                        html.Button('Reload Beta Plot', id='reload-beta-button', n_clicks=0),\n",
    "                                        dash_table.DataTable(\n",
    "                                                              id='beta-maf-table',\n",
    "                                                              columns=[{\"name\": i, \"id\": i} for i in beta_table_cols],\n",
    "                                                              data=blank_beta_df.to_dict('records'),\n",
    "                                                              filter_action=\"native\",\n",
    "                                                              sort_action=\"native\",\n",
    "                                                              sort_mode=\"multi\",\n",
    "                                                              column_selectable=\"single\",\n",
    "                                                              row_selectable=\"multi\",\n",
    "                                                              selected_columns=[],\n",
    "                                                              selected_rows=[0],\n",
    "                                                              page_action=\"native\",\n",
    "                                                              page_current= 0,\n",
    "                                                              page_size= 12,\n",
    "                                         ), \n",
    "                                        html.Div([html.P('Purity: ', style={'display': 'inline'}), html.P(0, id='beta-graph-purity', style={'display': 'inline'})]), \n",
    "                                        html.Div([html.P('Ploidy: ', style={'display': 'inline'}), html.P(0, id='beta-graph-ploidy', style={'display': 'inline'})]), \n",
    "                                        dcc.Graph(id='beta-graph', figure={})]), # todo just make name the heading\n",
    "                              callback_output=[Output('beta-graph', 'figure'), \n",
    "                                               Output('beta-maf-table', 'data'), \n",
    "                                               Output('beta-maf-table', 'selected_rows'),\n",
    "                                               Output('beta-graph-purity', 'children'),\n",
    "                                               Output('beta-graph-ploidy', 'children')\n",
    "                                              ],\n",
    "                              callback_input=[Input('reload-beta-button', 'n_clicks'), \n",
    "                                               State('beta-maf-table', 'selected_rows')],\n",
    "                              func=beta_graph_callback, \n",
    "                              add_autofill=True,\n",
    "                              autofill_dict={'purity': Input('beta-graph-purity', 'children')},\n",
    "                              beta_table_fn_col='BETA_annot_maf_fn',\n",
    "                              beta_table_display_col=beta_table_cols\n",
    "                             )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 349,
   "id": "7759a48e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dash app running on http://0.0.0.0:8052/\n"
     ]
    }
   ],
   "source": [
    "test_app.run_app(mode='external', port=8052)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "94e580e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['BETA_FLAG_not_enough_drivers',\n",
       " 'BETA_annot_maf_fn',\n",
       " 'BETA_clonal_muts',\n",
       " 'BETA_clonal_muts_genes',\n",
       " 'BETA_half_purity',\n",
       " 'BETA_has_beta_solution',\n",
       " 'BETA_num_clonal_drivers',\n",
       " 'BETA_ploidy',\n",
       " 'BETA_purity',\n",
       " 'BETA_purity_lower',\n",
       " 'BETA_purity_upper',\n",
       " 'BETA_ranked_maf_path',\n",
       " 'BETA_subclonal_muts',\n",
       " 'BETA_subclonal_muts_genes',\n",
       " 'ABSOLUTE_purity',\n",
       " 'ABSOLUTE_ploidy',\n",
       " 'ABSOLUTE_purity_upper',\n",
       " 'ABSOLUTE_purity_lower',\n",
       " 'ABSOLUTE_absolute_mod_tab_annotated_fn',\n",
       " 'ABSOLUTE_rdata_path',\n",
       " 'ABSOLUTE_FLAG_no_good_absolute_solution',\n",
       " 'AUTO_FINAL_method',\n",
       " 'AUTO_FINAL_purity',\n",
       " 'AUTO_FINAL_purity_upper',\n",
       " 'AUTO_FINAL_purity_lower',\n",
       " 'AUTO_FINAL_ploidy',\n",
       " 'AUTO_FINAL_FLAG_no_purity_call',\n",
       " 'BETA_annot_maf_gsurl',\n",
       " 'CGA_BIOSPECIMEN_db_id',\n",
       " 'CGA_BIOSPECIMEN_aliases',\n",
       " 'CGA_BIOSPECIMEN_participant_id',\n",
       " 'CGA_BIOSPECIMEN_irb_protocols',\n",
       " 'CGA_BIOSPECIMEN_collection_date_dfd',\n",
       " 'CGA_BIOSPECIMEN_tissue_site',\n",
       " 'CGA_BIOSPECIMEN_tissue_site_detail',\n",
       " 'CGA_BIOSPECIMEN_original_material_type',\n",
       " 'CGA_BIOSPECIMEN_submitted_material_type',\n",
       " 'CGA_BIOSPECIMEN_preservation_method',\n",
       " 'CGA_BIOSPECIMEN_primary_site',\n",
       " 'CGA_BIOSPECIMEN_tumor_normal',\n",
       " 'CGA_BIOSPECIMEN_tumor_state',\n",
       " 'CGA_BIOSPECIMEN_associated_samples',\n",
       " 'CGA_BIOSPECIMEN_sample_statuses',\n",
       " 'CGA_BIOSPECIMEN_sm_id',\n",
       " 'CGA_BIOSPECIMEN_gp_submitted_date',\n",
       " 'CGA_BIOSPECIMEN_notes',\n",
       " 'CGA_BIOSPECIMEN_post_allogeneic_transplant',\n",
       " 'CGA_BIOSPECIMEN_primary_cohort',\n",
       " 'CGA_BIOSPECIMEN_delete_biospecimen',\n",
       " 'CGA_PATIENT_db_id',\n",
       " 'CGA_PATIENT_aliases',\n",
       " 'CGA_PATIENT_tumor_primary_site',\n",
       " 'CGA_PATIENT_tumor_morphology',\n",
       " 'CGA_PATIENT_tumor_molecular_subtype',\n",
       " 'CGA_PATIENT_biospecimen_type_count',\n",
       " 'CGA_PATIENT_main_normal',\n",
       " 'CGA_PATIENT_gender',\n",
       " 'CGA_PATIENT_race',\n",
       " 'CGA_PATIENT_age_at_diagnosis',\n",
       " 'CGA_PATIENT_year_of_diagnosis',\n",
       " 'CGA_PATIENT_cancer_stage',\n",
       " 'CGA_PATIENT_vital_status',\n",
       " 'CGA_PATIENT_death_date_dfd',\n",
       " 'CGA_PATIENT_biospecimen_total_count',\n",
       " 'CGA_PATIENT_treatment_count',\n",
       " 'CGA_PATIENT_follow_up_date',\n",
       " 'CGA_PATIENT_notes',\n",
       " 'CGA_PATIENT_primary_cohort',\n",
       " 'CGA_PATIENT_all_cohorts',\n",
       " 'CGA_PATIENT_cohorts_to_add',\n",
       " 'CGA_PATIENT_delete_participant',\n",
       " 'CGA_participant_id',\n",
       " 'DFCI_local_sample_dfci_maf_fn',\n",
       " 'DFCI_bucket_sample_dfci_maf_fn',\n",
       " 'participant_id',\n",
       " 'DFCI_SAMPLE_Participant ID:',\n",
       " 'DFCI_SAMPLE_Gender:',\n",
       " 'DFCI_SAMPLE_Age at AML/MDS dx:',\n",
       " 'DFCI_SAMPLE_Status:',\n",
       " 'DFCI_SAMPLE_Sample Source:',\n",
       " 'DFCI_SAMPLE_Days from diagnosis:',\n",
       " 'DFCI_SAMPLE_Allo:',\n",
       " 'DFCI_SAMPLE_Anthracycline:',\n",
       " 'DFCI_SAMPLE_anti-CD33:',\n",
       " 'DFCI_SAMPLE_AraC:',\n",
       " 'DFCI_SAMPLE_BCR/ABL TKI:',\n",
       " 'DFCI_SAMPLE_Cladribine:',\n",
       " 'DFCI_SAMPLE_Clofarabine:',\n",
       " 'DFCI_SAMPLE_Cyclophosphamide:',\n",
       " 'DFCI_SAMPLE_Etoposide:',\n",
       " 'DFCI_SAMPLE_FLT3 TKI:',\n",
       " 'DFCI_SAMPLE_Fludarabine:',\n",
       " 'DFCI_SAMPLE_HMA:',\n",
       " 'DFCI_SAMPLE_Hydroxyurea:',\n",
       " 'DFCI_SAMPLE_IDHi:',\n",
       " 'DFCI_SAMPLE_Lenalidomide:',\n",
       " 'DFCI_SAMPLE_MTX:',\n",
       " 'DFCI_SAMPLE_Other:',\n",
       " 'DFCI_SAMPLE_Platinum:',\n",
       " 'DFCI_SAMPLE_Venetoclax:',\n",
       " 'DFCI_GENETICS_Participant ID:',\n",
       " 'DFCI_GENETICS_Gender:',\n",
       " 'DFCI_GENETICS_Molecular:',\n",
       " 'DFCI_GENETICS_DNMT3A',\n",
       " 'DFCI_GENETICS_TET2',\n",
       " 'DFCI_GENETICS_NPM1',\n",
       " 'DFCI_GENETICS_FLT3',\n",
       " 'DFCI_GENETICS_RAS',\n",
       " 'DFCI_GENETICS_TP53',\n",
       " 'DFCI_GENETICS_IDH1/2',\n",
       " 'DFCI_GENETICS_Other',\n",
       " 'DFCI_GENETICS_Splicing',\n",
       " 'DFCI_GENETICS_RUNX1',\n",
       " 'DFCI_GENETICS_ASXL1',\n",
       " 'DFCI_CYTO_Gender:',\n",
       " 'DFCI_CYTO_Cytogenetics:',\n",
       " 'DFCI_CYTO_Normal (1 = yes, 0 = no)',\n",
       " 'DFCI_CYTO_Diploid (46XX or 46XY with any number of other abnormalities)',\n",
       " 'DFCI_CYTO_Loss  Y or X',\n",
       " 'DFCI_CYTO_11q23 abnormality',\n",
       " 'DFCI_CYTO_Trisomy 8',\n",
       " 'DFCI_CYTO_Inv(3q) or t(3q)',\n",
       " 'DFCI_CYTO_del(20q)',\n",
       " 'DFCI_CYTO_del(12p)',\n",
       " 'DFCI_CYTO_Abnormality 7q/-7',\n",
       " 'DFCI_CYTO_Abnormality 5q/-5',\n",
       " 'DFCI_CYTO_Abnormality 17p/-17',\n",
       " 'DFCI_CYTO_Trisomy 21',\n",
       " 'DFCI_CYTO_Complex (3 or more)/monosomal',\n",
       " 'DFCI_CYTO_inv(16)/t(8;21)',\n",
       " 'DFCI_CYTO_t(9;22)',\n",
       " 'DFCI_CYTO_t(15;17)',\n",
       " 'DFCI_PRIOR_THERAPY_# of lines:',\n",
       " 'DFCI_PRIOR_THERAPY_XRT',\n",
       " 'DFCI_PRIOR_THERAPY_5-FU/Capecitabine',\n",
       " 'DFCI_PRIOR_THERAPY_6MP',\n",
       " 'DFCI_PRIOR_THERAPY_Abatacept',\n",
       " 'DFCI_PRIOR_THERAPY_Agent Orange',\n",
       " 'DFCI_PRIOR_THERAPY_allo-SCT',\n",
       " 'DFCI_PRIOR_THERAPY_Anthracycline',\n",
       " 'DFCI_PRIOR_THERAPY_anti-CD20',\n",
       " 'DFCI_PRIOR_THERAPY_anti-HER2',\n",
       " 'DFCI_PRIOR_THERAPY_anti-TNF',\n",
       " 'DFCI_PRIOR_THERAPY_Auto SCT (NOS):',\n",
       " 'DFCI_PRIOR_THERAPY_Azathioprine',\n",
       " 'DFCI_PRIOR_THERAPY_BCNU',\n",
       " 'DFCI_PRIOR_THERAPY_Bendamustine',\n",
       " 'DFCI_PRIOR_THERAPY_Bleomycin',\n",
       " 'DFCI_PRIOR_THERAPY_Brentuximab',\n",
       " 'DFCI_PRIOR_THERAPY_BTKi',\n",
       " 'DFCI_PRIOR_THERAPY_Busulfan',\n",
       " 'DFCI_PRIOR_THERAPY_CDK4/6i',\n",
       " 'DFCI_PRIOR_THERAPY_Chemo (NOS)',\n",
       " 'DFCI_PRIOR_THERAPY_Chlorambucil',\n",
       " 'DFCI_PRIOR_THERAPY_Cladribine',\n",
       " 'DFCI_PRIOR_THERAPY_Cyclophosphamide/Ifosphamide/Temozolamide',\n",
       " 'DFCI_PRIOR_THERAPY_Cyclosporine',\n",
       " 'DFCI_PRIOR_THERAPY_Cytarabine',\n",
       " 'DFCI_PRIOR_THERAPY_Dacarbazine/Procarbazine',\n",
       " 'DFCI_PRIOR_THERAPY_Enzastaurin',\n",
       " 'DFCI_PRIOR_THERAPY_Etoposide',\n",
       " 'DFCI_PRIOR_THERAPY_Fludarabine',\n",
       " 'DFCI_PRIOR_THERAPY_Gemcitabine',\n",
       " 'DFCI_PRIOR_THERAPY_HDACi',\n",
       " 'DFCI_PRIOR_THERAPY_Hormonal therapy',\n",
       " 'DFCI_PRIOR_THERAPY_Hydroxychloroquine',\n",
       " 'DFCI_PRIOR_THERAPY_Hydroxyurea',\n",
       " 'DFCI_PRIOR_THERAPY_Immunotherapy (Ipi/Nivo/Pembro)',\n",
       " 'DFCI_PRIOR_THERAPY_Mechlorethamine',\n",
       " 'DFCI_PRIOR_THERAPY_Melphalan',\n",
       " 'DFCI_PRIOR_THERAPY_Methotrexate',\n",
       " 'DFCI_PRIOR_THERAPY_Mitomycin-C',\n",
       " 'DFCI_PRIOR_THERAPY_mTORi',\n",
       " 'DFCI_PRIOR_THERAPY_PARPi',\n",
       " 'DFCI_PRIOR_THERAPY_Pemetrexed',\n",
       " 'DFCI_PRIOR_THERAPY_Pentostatin',\n",
       " 'DFCI_PRIOR_THERAPY_Platinum',\n",
       " 'DFCI_PRIOR_THERAPY_Steroid',\n",
       " 'DFCI_PRIOR_THERAPY_Sulfasalazine',\n",
       " 'DFCI_PRIOR_THERAPY_Targeted therapy NOS',\n",
       " 'DFCI_PRIOR_THERAPY_Taxane',\n",
       " 'DFCI_PRIOR_THERAPY_Thalidomide analog',\n",
       " 'DFCI_PRIOR_THERAPY_Thiotepa',\n",
       " 'DFCI_PRIOR_THERAPY_Topotecan/Irinotecan',\n",
       " 'DFCI_PRIOR_THERAPY_Proteasome inhibitor',\n",
       " 'DFCI_PRIOR_THERAPY_Venetoclax',\n",
       " 'DFCI_PRIOR_THERAPY_Vinca',\n",
       " 'DFCI_PRIOR_THERAPY_Vinorelbine/Navelbine',\n",
       " 'DFCI_CANCER_HISTORY_Dx:',\n",
       " 'DFCI_CANCER_HISTORY_Age at Dx:',\n",
       " 'DFCI_CANCER_HISTORY_Chemo:',\n",
       " 'DFCI_CANCER_HISTORY_Radiation:',\n",
       " 'DFCI_CANCER_HISTORY_Transplant (date):',\n",
       " 'DFCI_CANCER_HISTORY_Cancer status at leukemia dx (Remission/Relapsed):',\n",
       " 'DFCI_CANCER_HISTORY_Dx:.1',\n",
       " 'DFCI_CANCER_HISTORY_Age at Dx:.1',\n",
       " 'DFCI_CANCER_HISTORY_Chemo:.1',\n",
       " 'DFCI_CANCER_HISTORY_Radiation:.1',\n",
       " 'DFCI_CANCER_HISTORY_Transplant (date):.1',\n",
       " 'DFCI_CANCER_HISTORY_Cancer status at leukemia dx (Remission/Relapsed):.1',\n",
       " 'DFCI_CANCER_HISTORY_Dx:.2',\n",
       " 'DFCI_CANCER_HISTORY_Age at Dx:.2',\n",
       " 'DFCI_CANCER_HISTORY_Chemo:.2',\n",
       " 'DFCI_CANCER_HISTORY_Radiation:.2',\n",
       " 'DFCI_CANCER_HISTORY_Transplant (date):.2',\n",
       " 'DFCI_CANCER_HISTORY_Cancer status at leukemia dx (Remission/Relapsed):.2',\n",
       " 'DFCI_CANCER_HISTORY_Dx:.3',\n",
       " 'DFCI_CANCER_HISTORY_Age at Dx:.3',\n",
       " 'DFCI_CANCER_HISTORY_Chemo:.3',\n",
       " 'DFCI_CANCER_HISTORY_Radiation:.3',\n",
       " 'DFCI_CANCER_HISTORY_Transplant (date):.3',\n",
       " 'DFCI_CANCER_HISTORY_Cancer status at leukemia dx (Remission/Relapsed):.3',\n",
       " 'DFCI_SURVIVAL_Gender:',\n",
       " 'DFCI_SURVIVAL_Gender:.1',\n",
       " 'DFCI_SURVIVAL_Age at AML/MDS dx:',\n",
       " 'DFCI_SURVIVAL_Age at AML/MDS dx:.1',\n",
       " 'DFCI_SURVIVAL_Death or last known alive (days from MDS/AML diagnosis):',\n",
       " 'DFCI_SURVIVAL_Age at last known alive (years):',\n",
       " 'DFCI_SURVIVAL_Disease status (Relapse/Remission):',\n",
       " 'DFCI_SURVIVAL_Alive (y/n):',\n",
       " 'DFCI_SURVIVAL_Cause of Death:',\n",
       " 'DFCI_SURVIVAL_Diagnosis:',\n",
       " 'DFCI_SURVIVAL_Subtype:',\n",
       " 'DFCI_SURVIVAL_Sample Date (DfD):',\n",
       " 'DFCI_SURVIVAL_Status:',\n",
       " 'DFCI_SURVIVAL_ELN risk group (by cytogenetics, does not include FLT3/NPM1):',\n",
       " 'DFCI_SURVIVAL_IPSS-R cytogenetic risk group:',\n",
       " 'DFCI_SURVIVAL_Treatment prior to first sample:',\n",
       " 'DFCI_participant_id',\n",
       " 'gpdw_DNA_WES_icev1_read_groups',\n",
       " 'pdb_DNA_WES_icev1_bait_set_id',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_read_coverage_lane',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_lane_list',\n",
       " 'gpdw_DNA_WES_icev1_hs_pct_selected_bases',\n",
       " 'participant',\n",
       " 'pdb_last_modified',\n",
       " 'unmatched_WES_OUTPUT_fracContam',\n",
       " 'unmatched_WES_OUTPUT_copy_number_qc_report_png',\n",
       " 'unmatched_WES_OUTPUT_number_mutations_rejected',\n",
       " 'unmatched_WES_OUTPUT_alleliccapseg_tsv',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_converted_oxog_metrics',\n",
       " 'unmatched_WES_OUTPUT_deTiN_number_added_SSNVs',\n",
       " 'unmatched_WES_OUTPUT_contamination_data',\n",
       " 'unmatched_WES_OUTPUT_cross_check_fingprt_min_lod_lanes',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_bait_bias_summary_metrics',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_pre_adapter_detail_metrics',\n",
       " 'unmatched_WES_OUTPUT_MuTect1_merged_coverage_wig',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_bait_bias_summary_metrics',\n",
       " 'unmatched_WES_OUTPUT_filter_passed_and_rejected_mutations',\n",
       " 'unmatched_WES_OUTPUT_STRELKA_VEP_annotated_vcf',\n",
       " 'unmatched_WES_OUTPUT_deTiN_indels',\n",
       " 'gpdw_DNA_WES_icev1_hs_pct_pf_reads',\n",
       " 'unmatched_WES_OUTPUT_strelka_all_indels',\n",
       " 'pdb_DNA_WES_icev1_gp_product_order',\n",
       " 'unmatched_WES_OUTPUT_contest_validation',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_quality_by_cycle',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_gc_bias_summary_metrics',\n",
       " 'gpdw_DNA_WES_icev1_reference_sequence_name',\n",
       " 'unmatched_WES_OUTPUT_filter_passed_mutations',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_quality_distribution_metrics',\n",
       " 'unmatched_WES_OUTPUT_deTiN_aSCNA_model',\n",
       " 'unmatched_WES_OUTPUT_gatk_cnv_all_plots',\n",
       " 'unmatched_WES_OUTPUT_gatk_cnv_pre_tn_coverage',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_hybrid_selection_metrics',\n",
       " 'unmatched_WES_OUTPUT_lego_plotter_figs',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_base_distribution_by_cycle',\n",
       " 'gpdw_DNA_WES_icev1_data_type',\n",
       " 'unmatched_WES_OUTPUT_blat_all_maf',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_hybrid_selection_metrics',\n",
       " 'unmatched_WES_OUTPUT_oxoG_OBF_number_mutations_passed',\n",
       " 'unmatched_WES_OUTPUT_alleliccapseg_skew',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_lane_list',\n",
       " 'unmatched_WES_OUTPUT_oxoG_OBF_passed_and_rejected_mutations',\n",
       " 'unmatched_WES_OUTPUT_copy_number_qc_report',\n",
       " 'gpdw_DNA_WES_icev1_hs_pct_usable_bases_on_target',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_insert_size_histogram',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_base_distribution_by_cycle_metrics',\n",
       " 'gpdw_DNA_WES_icev1_hs_on_target_bases',\n",
       " 'unmatched_WES_OUTPUT_deTiN_TiN_modes_plot',\n",
       " 'unmatched_WES_OUTPUT_blat_passed_mutations',\n",
       " 'pdb_post_allogenic_transplant',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_insert_size_metrics',\n",
       " 'unmatched_WES_OUTPUT_deTiN_call_stats',\n",
       " 'unmatched_WES_OUTPUT_MUTECT2_VCF_ALL',\n",
       " 'pdb_DNA_WES_icev1_last_modified',\n",
       " 'gpdw_DNA_WES_icev1_mean_depth',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_pre_adapter_summary_metrics',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_pre_adapter_detail_metrics',\n",
       " 'gpdw_DNA_WES_icev1_hs_target_territory',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_alignment_summary_metrics',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_quality_distribution',\n",
       " 'unmatched_WES_OUTPUT_ffpe_OBF_number_mutations_rejected',\n",
       " 'gpdw_DNA_WES_icev1_release_date',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_converted_oxog_metrics',\n",
       " 'unmatched_WES_OUTPUT_ffpe_OBF_passed_and_rejected_mutations',\n",
       " 'pdb_tumor_normal',\n",
       " 'unmatched_WES_OUTPUT_contest_base_report',\n",
       " 'unmatched_WES_OUTPUT_absolute_rdata',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_quality_by_cycle_metrics',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_alignment_summary_metrics',\n",
       " 'unmatched_WES_OUTPUT_cross_check_fingprt_metrics',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_quality_by_cycle_metrics',\n",
       " 'gpdw_DNA_WES_icev1_mean_read_length',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_gc_bias',\n",
       " 'unmatched_WES_OUTPUT_lego_plotter_pngs',\n",
       " 'gpdw_DNA_WES_icev1_read_pairs',\n",
       " 'gpdw_DNA_WES_icev1_median_insert_size',\n",
       " 'gpdw_DNA_WES_icev1_hs_mean_bait_coverage',\n",
       " 'gpdw_DNA_WES_icev1_bait_set',\n",
       " 'unmatched_WES_OUTPUT_TiN_CI',\n",
       " 'unmatched_WES_OUTPUT_alleliccapseg_plot',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_gc_bias',\n",
       " 'unmatched_WES_OUTPUT_MUTECT2_VEP_annotated_vcf',\n",
       " 'unmatched_WES_OUTPUT_lego_plotter_pss',\n",
       " 'unmatched_WES_OUTPUT_mutect_force_call_pw',\n",
       " 'unmatched_WES_OUTPUT_blat_number_mutations_rejected',\n",
       " 'pdb_original_material_type',\n",
       " 'unmatched_WES_OUTPUT_lego_plotter_ais',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_bait_bias_detail_metrics',\n",
       " 'pdb_submitted_material_type',\n",
       " 'pdb_preservation_method',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_quality_distribution_metrics',\n",
       " 'pdb_collection_date_dfd',\n",
       " 'pdb_tumor_state',\n",
       " 'unmatched_WES_OUTPUT_strelka_passed_indels',\n",
       " 'unmatched_WES_OUTPUT_deTiN_segments',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_bam_validation',\n",
       " 'unmatched_WES_OUTPUT_mutation_validator_validated_maf',\n",
       " 'unmatched_WES_OUTPUT_deTiN_aSCNA_scatter_plot',\n",
       " 'unmatched_WES_OUTPUT_gatk_cnv_tn_coverage',\n",
       " 'gpdw_DNA_WES_icev1_sample_type',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_gc_bias_summary_metrics',\n",
       " 'unmatched_WES_OUTPUT_strelka_all_snvs',\n",
       " 'gpdw_DNA_WES_icev1_sample_alias',\n",
       " 'unmatched_WES_OUTPUT_oxoG_OBF_passed_mutations',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_base_distribution_by_cycle_metrics',\n",
       " 'gpdw_DNA_WES_icev1_min_insert_size',\n",
       " 'unmatched_WES_OUTPUT_gatk_het_ad_normal',\n",
       " 'unmatched_WES_OUTPUT_MUTECT1_CS_SNV',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_bam_validation',\n",
       " 'unmatched_WES_OUTPUT_gatk_cnv_seg_file',\n",
       " 'unmatched_WES_OUTPUT_ffpe_OBF_figures',\n",
       " 'unmatched_WES_OUTPUT_mutect1_snv_mutect2_indel_strelka_indel_annotated_maf',\n",
       " 'pdb_DNA_WES_icev1_delivery_workspace',\n",
       " 'gpdw_DNA_WES_icev1_hs_median_target_coverage',\n",
       " 'unmatched_WES_OUTPUT_oxoG_OBF_number_mutations_rejected',\n",
       " 'gpdw_DNA_WES_icev1_crai_or_bai_path',\n",
       " 'unmatched_WES_OUTPUT_TiN',\n",
       " 'unmatched_WES_OUTPUT_mut_legos_html',\n",
       " 'unmatched_WES_OUTPUT_absolute_highres_plot',\n",
       " 'gpdw_DNA_WES_icev1_aligned_reads',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_insert_size_histogram',\n",
       " 'unmatched_WES_OUTPUT_number_mutations_passed',\n",
       " 'gpdw_DNA_WES_icev1_hs_pf_unique_reads',\n",
       " 'unmatched_WES_OUTPUT_ffpe_OBF_number_mutations_passed',\n",
       " 'pdb_gp_submitted_date',\n",
       " 'unmatched_WES_OUTPUT_ffpe_OBF_passed_mutations',\n",
       " 'unmatched_WES_OUTPUT_filters_passed_merged_intersection_maf',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_read_coverage_lane',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_quality_yield_metrics',\n",
       " 'gpdw_DNA_WES_icev1_hs_mean_target_coverage',\n",
       " 'unmatched_WES_OUTPUT_MUTECT1_VEP_annotated_vcf',\n",
       " 'unmatched_WES_OUTPUT_copy_number_qc_mix_ups',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_insert_size_metrics',\n",
       " 'unmatched_WES_OUTPUT_contestAFFile',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_quality_by_cycle',\n",
       " 'gpdw_DNA_WES_icev1_sequencing_project',\n",
       " 'unmatched_WES_OUTPUT_MuTect1_merged_power_wig',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_base_distribution_by_cycle',\n",
       " 'gpdw_DNA_WES_icev1_total_reads',\n",
       " 'unmatched_WES_OUTPUT_deTiN_SSNVs_plot',\n",
       " 'unmatched_WES_OUTPUT_gatk_het_ad_tumor',\n",
       " 'unmatched_WES_OUTPUT_cross_check_fingprt_report',\n",
       " 'unmatched_WES_OUTPUT_filters_passed_merged_union_maf',\n",
       " 'unmatched_WES_OUTPUT_gatk_cnv_coverage_file',\n",
       " 'unmatched_WES_OUTPUT_MUTECT2_VCF_INDELS',\n",
       " 'unmatched_WES_OUTPUT_blat_number_mutations_passed',\n",
       " 'unmatched_WES_OUTPUT_ffpe_OBF_q_val',\n",
       " 'unmatched_WES_OUTPUT_cross_check_fingprt_min_lod_value',\n",
       " 'gpdw_DNA_WES_icev1_cram_or_bam_path',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_bait_bias_detail_metrics',\n",
       " 'gpdw_DNA_WES_icev1_reads_aligned_in_pairs',\n",
       " 'unmatched_WES_OUTPUT_mutect_force_call_cs',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_pre_adapter_summary_metrics',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_quality_yield_metrics',\n",
       " 'unmatched_WES_OUTPUT_deTiN_aSCNA_kmeans_RSS_plot',\n",
       " 'unmatched_WES_OUTPUT_tumor_bam_gc_bias_detail_metrics',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_gc_bias_detail_metrics',\n",
       " 'unmatched_WES_OUTPUT_oxoG_OBF_q_val',\n",
       " 'unmatched_WES_OUTPUT_normal_bam_quality_distribution',\n",
       " 'unmatched_WES_OUTPUT_strelka_passed_snvs',\n",
       " 'pdb_DNA_WES_icev1_submitted_name',\n",
       " 'unmatched_WES_OUTPUT_mutect_force_call_cw',\n",
       " 'unmatched_WES_OUTPUT_oxoG_OBF_figures',\n",
       " 'gpdw_RNA_RNASeq_mean_depth',\n",
       " 'RNA_OUTPUT_rnaseqc2_exon_counts',\n",
       " 'pdb_RNA_RNASeq_last_modified',\n",
       " 'RNA_OUTPUT_star_junctions_pass1',\n",
       " 'RNA_OUTPUT_star_md_bam_index',\n",
       " 'gpdw_RNA_RNASeq_sample_type',\n",
       " 'RNA_OUTPUT_rnaseqc2_metrics',\n",
       " 'pdb_RNA_RNASeq_submitted_name',\n",
       " 'gpdw_RNA_RNASeq_cram_or_bam_path',\n",
       " 'RNA_OUTPUT_marked_dup_metrics',\n",
       " 'RNA_OUTPUT_star_md_bam_file',\n",
       " 'RNA_OUTPUT_rnaseqc2_gc_content',\n",
       " 'RNA_OUTPUT_rnaseqc2_gene_counts',\n",
       " 'RNA_OUTPUT_star_bam_index',\n",
       " 'gpdw_RNA_RNASeq_mean_read_length',\n",
       " 'RNA_OUTPUT_fastq2',\n",
       " 'gpdw_RNA_RNASeq_read_groups',\n",
       " 'gpdw_RNA_RNASeq_release_date',\n",
       " 'RNA_OUTPUT_star_read_counts',\n",
       " 'gpdw_RNA_RNASeq_median_insert_size',\n",
       " 'RNA_OUTPUT_star_logs',\n",
       " 'RNA_OUTPUT_star_chimeric_bam_file',\n",
       " 'RNA_OUTPUT_rnaseqc2_gene_tpm',\n",
       " 'gpdw_RNA_RNASeq_sequencing_project',\n",
       " 'gpdw_RNA_RNASeq_sample_alias',\n",
       " 'pdb_RNA_RNASeq_gp_product_order',\n",
       " 'gpdw_RNA_RNASeq_aligned_reads',\n",
       " 'gpdw_RNA_RNASeq_min_insert_size',\n",
       " 'pdb_RNA_RNASeq_delivery_workspace',\n",
       " 'RNA_OUTPUT_star_transcriptome_bam',\n",
       " 'RNA_OUTPUT_star_junctions',\n",
       " 'RNA_OUTPUT_star_chimeric_bam_index',\n",
       " 'RNA_OUTPUT_rsem_isoforms',\n",
       " 'RNA_OUTPUT_fastq1',\n",
       " 'gpdw_RNA_RNASeq_crai_or_bai_path',\n",
       " 'RNA_OUTPUT_star_bam_file',\n",
       " 'RNA_OUTPUT_star_chimeric_junctions',\n",
       " 'RNA_OUTPUT_rsem_genes',\n",
       " 'RNA_OUTPUT_rnaseqc2_insertsize_distr',\n",
       " 'pdb_sm_id',\n",
       " 'CONSOLE_unmatched_WES_OUTPUT_absolute_highres_plot',\n",
       " 'CONSOLE_unmatched_WES_OUTPUT_alleliccapseg_plot',\n",
       " 'manual_purity',\n",
       " 'manual_purity_lower',\n",
       " 'manual_purity_upper',\n",
       " 'manual_ploidy',\n",
       " 'manual_confidence',\n",
       " 'manual_flags',\n",
       " 'last_manual_update',\n",
       " 'manual_method',\n",
       " 'MAFLITE',\n",
       " 'VCF']"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'sample-info-component': [], 'maf-component-id': [], 'beta-graph': [0, [0]]}\n",
      "{'sample-info-component': [], 'maf-component-id': [], 'beta-graph': [0, [0]]}\n",
      "prop_id: APP-dropdown-data-state\n",
      "0\n",
      "sample-info-component\n",
      "\n",
      "maf-component-id\n",
      "\n",
      "beta-graph\n",
      "0 [0]\n"
     ]
    }
   ],
   "source": [
    "test_rd.data.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b1a06a4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
